---
title: "03-Adult-data-analysis-QuantSeq2020"
author: "Laura H Spencer"
date: "1/24/2021"
output: html_document
---

```{r}
getwd()
```

### Load libraries and source scripts 

```{r, message=FALSE, warning=FALSE, results=FALSE}
source("../references/biostats.R")

list.of.packages <- c("DESeq2", "RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "cowplot", "clipr", "janitor", "ggpubr", "forcats", "apeglm") #add new libraries here 
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})
```

### Inspect total counts by sample for adult samples 
```{r}
load(file = "../results/gene-counts-filtered") #object = counts.filtered
load(file="../raw-data/key.filtered") #object = key

adult.samples <- key.filtered %>% filter(stage=="adult") %>% dplyr::select(sample) %>% unlist() %>% as.vector() %>% as.character()

ggplotly(
ggplot(data = counts.filtered %>%
         dplyr::select(adult.samples) %>%
         colSums() %>% data.frame() %>%
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample"))  +
           geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample, adult ctenidia tissue") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()))
```

### Load transformed counts object

```{r}
load(file = "../results/gene-counts-trans") #object = counts.t
```

### Extract adult samples only
```{r}
counts.t.adult <- as.data.frame(counts.t[adult.samples, ])
counts.t.adult[,1:3] #take a peak at the resulting dataframe 
```
## Optional 

### Pre-filtering - remove rows (genes) with less than a total of 10 reads (across all samples) (why?)

```{r}
keep <- colSums(counts.t.adult) >= 10
counts.ts.adult <- counts.t.adult[,keep]

print(paste("# genes remaining after pre-filtering:", ncol(counts.ts.adult)))
print(paste("# of genes dropped:", ncol(counts.t.adult) - ncol(counts.ts.adult), sep=" "))

# How many genes per sample? 
data.frame(rowSums(counts.ts.adult != 0)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample") %>% 
  summarise(mean=mean(count.total), sd=sd(count.total), se=sd/sqrt(length(sample)))

# How many reads per sample? 
data.frame(rowSums(counts.ts.adult)) %>% 
                  dplyr::rename(read.total = 1) %>% 
                  rownames_to_column(var="sample") %>%  
summarise(min=min(read.total), max=max(read.total), mean=mean(read.total), 
          sd=sd(read.total), se=sd/sqrt(length(sample))) #use this to average across all samples 
```

### How many genes were identified in each sample?

```{r}
ggplotly(
ggplot(data = data.frame(rowSums(counts.t.adult != 0)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample")) +
           geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total # genes by sample, adult ctenidia tissue") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```


## Use foa.plots to visualize data a bit: 

### In how many samples does each gene occur? 
  - The **first four plots portray the gene’ frequency of occurrence among samples** in a number of different ways – either as an empirical cumulative distribution function (ECDF) of gene occurrence or as a histogram of gene occurrence. 

### What is the mean abundance of each gene when it occurs (not averaging zeros for samples where it is absent)? 
  - The **fifth plot is an ECDF of gene mean abundance.** X-axis is samples, ranked from 1-n in terms of mean gene abundance. 
  
### Is the mean abundance of genes correlated with the number of samples they occur in? 
  - The **sixth plot is a scatter plot of frequency of occurrence against mean abundance**. Is there any apparent relationship between the two? Are the widespread genes also generally more abundant? Are there many widespread genes that occur at low abundance? Conversely, are there genes much less widespread, but abundant where they occur?

### Is the total abundance of gene in a sample correlated with the number of gene in a sample? To answer this question, first it is instructive to look at the number of gene per sample. 
  - The **eighth plot depicts the ECDF of sample richness.** Are there any interesting patterns? For example, do most samples support an average number of gene, while only a few samples supporting either very few or very many gene? Or is the pattern different?

### Second, what is the pattern in the distribution of sample total abundance? 
  - The **ninth plot is the ECDF of total sample abundance.** How does it compare to the ECDF of sample richness?

### Finally, to answer the question on the relation between total abundance and number of gene/sample ...
  - The **last plot is a scatter plot of the two variables.** Is there is relationship between the _number of genes per sample and the total abundance?_ Do gene-rich samples generally have a greater total abundance of those genes as well? 

```{r}
# note: you'll need to press return in the console for all plots 
#foa.plots(counts.ts.adult)
```

### Merge sample key info to count data, then sort, and generate heat map for initial inspection by treatment 
```{r}
# merge count data with sample key, reset row names as sample names, and arrange by infection, then temperature, then day 
counts.tsk.adult <- merge(x=key, by.x="sample", y=counts.ts.adult, by.y="row.names") %>% 
  arrange(stage, population, pCO2.parent)  %>% column_to_rownames(var="sample") 

head(counts.tsk.adult[1:24]) #check out results of merge/arrange
#counts.tsk.adult %>% dplyr::select(starts_with("OLUR")) #this is code to get only the gene columns 
```

### Generate heat map of counts before DESeq processing / analysis 

NOTE: scale="column" b/c range of counts is so huge, so counts have been scaled 

```{r}
# pheatmap(data.matrix(counts.tsk.adult %>% dplyr::select(starts_with("OLUR"))), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, 
#                      show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE, 
#                      scale="column", color=c("dodgerblue3", "goldenrod1"), 
#                      main = "Oly adult ctendia gene counts", annotation_row=counts.tsk.adult[,c("population", "pCO2.parent")],
#          filename = "../results/heatmap-adult-counts.png")
```

Resulting heat map is located in results/ directory: 
![heatmap-adult-counts.png](https://github.com/laurahspencer/O.lurida_QuantSeq-2020/blob/master/results/heatmap-adult-counts.png?raw=true)

# Analysis in DESeq2  

### Reformat for DESeq, ensure correct sample order for 

NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

```{r}
all(rownames(counts.tsk.adult) == counts.tsk.adult %>% dplyr::select(starts_with("OLUR")) %>% t() %>% colnames()) #check that rownames of untransformed matrix match column names of transformed matrix. Should print 'TRUE' 
```

# Generate DESeq datasets with various treatment comparisons  

```{r}
#counts.DESeq <- counts.tsk.adult[-which(rownames(counts.tsk.adult) %in% "571_larvae"), grepl("OLUR", colnames(counts.tsk.adult))] %>% t()
#key.DESeq <- counts.tsk.adult[-which(rownames(counts.tsk.adult) %in% "571_larvae"),c("population", "pCO2.parent")] 

dds.adult <- DESeqDataSetFromMatrix(countData = counts.tsk.adult[,grepl("OLUR", colnames(counts.tsk.adult))] %>% t(),
                              colData = counts.tsk.adult[,c("population", "pCO2.parent")] ,
                              design = ~ population + pCO2.parent)
```

# Visualize data via PCAs and heat maps 

# HERE: REDO PCA USING CODE FROM MULTIVARIATE STATS 

## Transform data 

- Here we transform counts using a variance stabilizing transformation (VST), since we have many samples. 
- Here we use `blind=FALSE` b/c we are interested in differences explained by experimental design, and may wish to use this transformed data in downstream analyses. 
```{r}
vsd.adult <- varianceStabilizingTransformation(dds.adult, blind=FALSE)
```

## Visualize sample clustering via PCA (after transformation)

NOTE: Hover over points to see the sample numbers

```{r}
# PCA with points color coded by population 
ggplotly(
  plotPCA(vsd.adult, intgroup="population") + 
           ggtitle("PCA by population (var-stabilizing transformed)") + 
    geom_point(size=3, aes(text=colnames(vsd.adult))) + 
    theme_minimal()+ stat_ellipse(), tooltip = "text")

# PCA with points color coded by parental pCO2 exposure 
ggplotly(
  plotPCA(vsd.adult, intgroup="pCO2.parent") + 
           ggtitle("PCA by parental pCO2 exposure (var-stabilizing transformed)") + geom_point(size=3, aes(text=colnames(vsd.adult))) + stat_ellipse() + theme_minimal(), tooltip = "text")


# PCA with points color coded by population and pco2 factors 
ggplotly(
  plotPCA(vsd.adult, intgroup=c("population","pCO2.parent")) + 
           ggtitle("PCA by population + parental pCO2 (var-stabilizing transformed)") + geom_point(size=3, aes(text=colnames(vsd.adult)))+ stat_ellipse(), tooltip = "text")

# PCA with same 2 colors for all populations (for prelim. results presentation)
ggplotly(
  plotPCA(vsd.adult, intgroup=c("population","pCO2.parent")) + 
           ggtitle("PCA by population + parental pCO2 (var-stabilizing transformed)") + geom_point(size=4, aes(text=colnames(vsd.adult))) + 
    scale_color_manual(values = c("#F8766D","#00BFC4","#F8766D","#00BFC4","#F8766D","#00BFC4")), tooltip = "text")

PCA.data.adult <- plotPCA(vsd.adult, intgroup=c("population","pCO2.parent"), returnData=TRUE)
save(PCA.data.adult, file="../results/PCA.data.adult")


ggplotly(
  plotPCA(vsd.adult, intgroup="pCO2.parent") + 
           ggtitle("my title") + 
    geom_point(size=3, aes(text=colnames(vsd.adult))) + stat_ellipse() + theme_minimal(), tooltip = "text")
```

### Generate heat maps before & after transformation  

```{r}
# extract treatment info from VSD transformation 
#vsd.adult.df <- as.data.frame(colData(vsd.adult)[,c("population", "pCO2.parent")])

# generate heatmap from untransformed counts 
#pheatmap(counts(dds.adult), cluster_rows=FALSE, show_rownames=FALSE,
#         cluster_cols=T, annotation_col=vsd.adult.df, scale = "row", main="QuantSeq, untransformed data (but scaled by rows")

# generate heatmap from VSD counts 
#pheatmap(assay(vsd.adult), cluster_rows=FALSE, show_rownames=FALSE,
#         cluster_cols=T, annotation_col=vsd.adult.df, main = "QuantSeq, VSD-transformed")
```

### Heatmap of the sample-to-sample distances
Another use of the transformed data is sample clustering. Here, we apply the dist function to the transpose of the transformed count matrix to get sample-to-sample distances.

A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.

```{r}
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

sampleDists.adults <- dist(t(assay(vsd.adult)))
sampleDistMatrix.adults <- as.matrix(sampleDists.adults)

# Here we show pCO2.parent + population 
rownames(sampleDistMatrix.adults) <- paste(vsd.adult$population, vsd.adult$pCO2.parent, sep="-") #set row names 
colnames(sampleDistMatrix.adults) <- NULL
pheatmap(sampleDistMatrix.adults,
         clustering_distance_rows=sampleDists.adults,
         clustering_distance_cols=sampleDists.adults,
         col=rev(colors), fontsize = 6)
```

## Differential Expression Analysis - multifactor design 

### Run the function `DESeq` to assess differential expression 

```{r}
dds.adult.DESeq <- DESeq(dds.adult) 
```

## Any DEGs between parental pCO2, all populations? 

```{r}
print("Comparison: parental pCO2 - All populations")
summary(res.adult.pco2 <- results(dds.adult.DESeq, contrast=c("pCO2.parent", "Ambient", "High"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pCO2, adult ctendia, comparison across all pops:",  sum(res.adult.pco2$padj < 0.05, na.rm=TRUE))
res.adult.pco2
(16+55)/(30978-4207)*100
```
Volcano Plots - comparing pCO2 exposure with all populations combined 

```{r}
plotMA(res.adult.pco2, ylim=c(-2,2))

resultsNames(dds.adult.DESeq)
resLFC <- lfcShrink(dds.adult.DESeq, coef="pCO2.parent_High_vs_Ambient", type="apeglm")
resNorm <- lfcShrink(dds.adult.DESeq, coef="pCO2.parent_High_vs_Ambient", type="normal")
resAsh <- lfcShrink(dds.adult.DESeq, coef="pCO2.parent_High_vs_Ambient", type="ashr")

require(ashr)
par(mfrow=c(2,2), mar=c(4,4,2,1))
plotMA(res.adult.pco2, ylim=c(-3.1,3), main="log2 FC / mean normalized counts\nby pCO2, all populations")
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal")
plotMA(resAsh, xlim=xlim, ylim=ylim, main="ashr")
```

Heatmap, DEGs among pCO2 exposure (not considering population)
Not super convincing! 
```{r}
diffex.adult <- subset(res.adult.pco2, padj < 0.05)
diffex.adult.counts <- subset(counts(dds.adult.DESeq), rownames(dds.adult.DESeq) %in% rownames(diffex.adult)) 
diffex.adult.counts <- diffex.adult.counts[,subset(key.filtered, stage=="adult")$sample]
dds.adult.df <- as.data.frame(colData(dds.adult)[,c("population", "pCO2.parent")])
dds.adult.df <- dds.adult.df[match(colnames(diffex.adult.counts), rownames(dds.adult.df)),]
all(colnames(diffex.adult.counts) == rownames(dds.adult.df)) #double check that samples are still in same order 

pheatmap(diffex.adult.counts[rownames(diffex.adult[order(diffex.adult$padj),]),], cluster_rows=F, 
         show_rownames=FALSE, cluster_columns=TRUE, na.rm=TRUE, 
         scale="row", main = "(Adults) DEGs by direct pCO2 exposure (all pops), sorted by p-value", 
         annotation_col=dds.adult.df, color=c("dodgerblue3", "goldenrod1"))

pheatmap(diffex.adult.counts, cluster_rows=T, show_rownames=F, cluster_columns=T, na.rm=TRUE, 
         scale="row", main = "(Adults) DEGs by direct pCO2 exposure, clustered", fontsize_col = 7, fontsize_row = 7, 
         annotation_col=dds.adult.df, color=c("dodgerblue3", "goldenrod1"))
```

# Plot counts for DEGs across all populations 

```{r}
adult.p05.names <-rownames(diffex.adult[order(diffex.adult$padj),])
count_plots = list()

#plot FB DEGs to find any 
for (i in 1:length(adult.p05.names)) {
a <-  plotCounts(dds.adult.DESeq, gene=adult.p05.names[i], intgroup=c("pCO2.parent"), returnData = TRUE)
b <- ggplot(a %>% rownames_to_column("sample"),
       aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
  geom_point(position=position_jitter(w = 0.1,h = 0)) +
    geom_text() +
    theme_bw() +
    ggtitle(adult.p05.names[i]) +
    theme(plot.title = element_text(hjust = 0.5))
count_plots[[i]] <- b
}
count_plots
```

```{r}
p <- plotPCA(vsd.adult[adult.p05.names,], intgroup=c("pCO2.parent", "population"), returnData=T)

p <- plotPCA(vsd.adult, intgroup=c("pCO2.parent", "population"), returnData=T)
ggplot(p, aes(x=PC1, y=PC2, group=pCO2.parent, colour=pCO2.parent, shape=population)) + geom_point(size=4, alpha=0.8) + 
  theme_minimal() + stat_ellipse() + ggtitle(expression(paste("PCA of all gene counts in adults by treatment & population"))) +
  xlab("PC1: 14% variance") + ylab("PC2: 10% variance") +
  scale_shape_manual(name="Cohort", values=c(15,17,16),labels=c("Dabob Bay"="Hood Canal", "Fidalgo Bay"="North Sound", "Oyster Bay C1"="South Sound")) + 
  scale_color_manual(name="Treatment", labels=c("Ambient"="Control", "High"="Acidification"),values=c("#67a9cf","#ef8a62"))

```

# Enrichment analysis, DEGs among pCO2 across all populations 

Enrichment Analysis! 

```{r}
# DEGs between pCO2 
diffex.adult %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID, Note), by = c("gene"="Name")) %>% 
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# All genes (background)
res.adult.pco2 %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID), by = c("gene"="Name")) %>% 
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

| Term | Count | Fold Enrichment | FDR |
|-|-|-|-|
| GO:0055114~oxidation-reduction process | 16 | 7.798607888631091 | 8.243364733269016E-8 |
| GO:0006805~xenobiotic metabolic process | 3 | 27.401086956521738 | 0.3923019577707598 |
| GO:0098869~cellular oxidant detoxification | 3 | 21.007499999999997 | 0.3923019577707598 |
| GO:0006749~glutathione metabolic process | 3 | 19.097727272727273 | 0.3923019577707598 |
| GO:0097267~omega-hydroxylase P450 pathway | 2 | 140.05 | 0.3923019577707598 |
| GO:0060315~negative regulation of ryanodine-sensitive calcium-release channel activity | 2 | 140.05 | 0.3923019577707598 |
| GO:0060316~positive regulation of ryanodine-sensitive calcium-release channel activity | 2 | 140.05 | 0.3923019577707598 |
| GO:0002489~antigen processing and presentation of endogenous peptide antigen via MHC class Ib via ER pathway, TAP-dependent | 2 | 105.0375 | 0.3923019577707598 |
| GO:0002485~antigen processing and presentation of endogenous peptide antigen via MHC class I via ER pathway, TAP-dependent | 2 | 105.0375 | 0.3923019577707598 |
| GO:0002591~positive regulation of antigen processing and presentation of peptide antigen via MHC class I | 2 | 105.0375 | 0.3923019577707598 |
| GO:0002481~antigen processing and presentation of exogenous protein antigen via MHC class Ib, TAP-dependent | 2 | 105.0375 | 0.3923019577707598 |
| GO:0055085~transmembrane transport | 5 | 4.752828054298643 | 0.3923019577707598 |
| GO:0008202~steroid metabolic process | 3 | 13.700543478260869 | 0.3923019577707598 |
| GO:0006869~lipid transport | 3 | 13.40904255319149 | 0.3923019577707598 |
| GO:0006629~lipid metabolic process | 4 | 6.513953488372094 | 0.3923019577707598 |
| GO:0071243~cellular response to arsenic-containing substance | 2 | 84.03 | 0.3923019577707598 |
| GO:0010880~regulation of release of sequestered calcium ion into cytosol by sarcoplasmic reticulum | 2 | 84.03 | 0.3923019577707598 |
| GO:0042178~xenobiotic catabolic process | 2 | 70.025 | 0.4202584895201619 |
| GO:0019852~L-ascorbic acid metabolic process | 2 | 70.025 | 0.4202584895201619 |
| GO:0009812~flavonoid metabolic process | 2 | 52.51875 | 0.5299315105884214 |
| GO:0045909~positive regulation of vasodilation | 2 | 46.68333333333334 | 0.5665058796300599 |
| GO:0042446~hormone biosynthetic process | 2 | 42.015 | 0.5734235070083432 |
| GO:0045332~phospholipid translocation | 2 | 42.015 | 0.5734235070083432 |
| GO:0042403~thyroid hormone metabolic process | 2 | 38.195454545454545 | 0.5997213086754268 |
| GO:0008152~metabolic process | 4 | 4.617032967032967 | 0.5997213086754268 |
| GO:0006091~generation of precursor metabolites and energy | 2 | 35.0125 | 0.605978858881677 |
| GO:0006855~drug transmembrane transport | 2 | 32.31923076923077 | 0.6307442328553533 |
| GO:0006810~transport | 6 | 2.7048283261802575 | 0.6309987973836262 |
| GO:0008544~epidermis development | 2 | 30.010714285714286 | 0.6309987973836262 |
| GO:0051923~sulfation | 2 | 28.01 | 0.6520696101522573 |
| GO:0019373~epoxygenase P450 pathway | 2 | 24.714705882352945 | 0.711973967729552 |
| GO:0055088~lipid homeostasis | 2 | 20.00714285714286 | 0.8188345087017934 |
| GO:0019369~arachidonic acid metabolic process | 2 | 20.00714285714286 | 0.8188345087017934 |

Read in GO terms 

```{r}
load(file="../references/GOslim")

# For REVIGO 
read_delim(file="../results/Adult_allpops_DEGs_EnrichedBP.txt", delim = "\t") %>% 
  mutate(GO = str_extract(Term, "GO(.*?)~")) %>% 
  mutate(GO = gsub("~", "", GO)) %>% dplyr::select(GO, PValue) %>% na.omit() %>% write_clip()
```


## Now do comparisons of genes by parental pCO2 within populations: 

```{r}
# Now add factor that defines interaction group 
dds.adult.DESeq$group <- factor(paste0(dds.adult.DESeq$population, dds.adult.DESeq$pCO2.parent))
design(dds.adult.DESeq) <- ~ group 
dds.adult.grp.DESeq <- DESeq(dds.adult.DESeq)
```
# 

```{r}
# Check Cooks distances to see if one sample is consistently higher than others (i.e. has many outliers)
par(mar=c(8,5,2,2))
boxplot(log10(assays(dds.adult.grp.DESeq)[["cooks"]]), range=0, las=2)
# looks pretty good  

# plot dispersion estimates 
plotDispEsts(dds.adult.grp.DESeq) #looks like the example in the tutorial - not worried about dispersion
```


## Extract differential expression results / comparisons among different factors 

Create results objects, but summary of results are also shown

NOTE: can only compare two treatments at a time

```{r}
# Here are all the possible contrasts I can make 
levels(dds.adult.grp.DESeq$group)

print("Comparison: parental pCO2 - Fidalgo Bay")
summary(res.adult.FB <- results(dds.adult.grp.DESeq, contrast=c("group", "Fidalgo BayHigh", "Fidalgo BayAmbient"), alpha=0.05))

print("Comparison: parental pCO2 - Dabob Bay")
summary(res.adult.DB <- results(dds.adult.grp.DESeq, contrast=c("group", "Dabob BayHigh", "Dabob BayAmbient"), alpha=0.05))

print("Comparison: parental pCO2 - Oyster Bay Cohort 1")
summary(res.adult.OB1 <- results(dds.adult.grp.DESeq, contrast=c("group", "Oyster Bay C1High", "Oyster Bay C1Ambient"), alpha=0.05))
```

Volcano Plots - comparing pCO2 exposure within each 

```{r}
plotMA(res.adult.FB, ylim=c(-4.6,5.5), main="log2 FC / mean normalized counts\nby pCO2, Fidalgo Bay")
plotMA(res.adult.DB, ylim=c(-4.6,5.7), main="log2 FC / mean normalized counts\nby pCO2, Dabob Bay")
plotMA(res.adult.OB1, ylim=c(-4,4.5), main="log2 FC / mean normalized counts\nby pCO2, Oyster Bay")
```


## Count # of genes diff expressed  (p-value <0.05) in each comparison 

```{r}
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Fidalgo Bay adult ctenidia:",  sum(res.adult.FB$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Dabob Bay adult ctenidia:",  sum(res.adult.DB$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Oyster Bay Cohort1 adult ctenidia:",  sum(res.adult.OB1$padj < 0.05, na.rm=TRUE))
```

## Extract stats for significantly different genes for each comparison 

```{r}
diffex.adult.FB <- subset(res.adult.FB, padj < 0.05)
diffex.adult.DB <- subset(res.adult.DB, padj < 0.05)
diffex.adult.OB1 <- subset(res.adult.OB1, padj < 0.05)
```

## Extract counts for differentially expressed genes for each comparison 

```{r}
diffex.adult.FB.counts <- subset(counts(dds.adult.grp.DESeq), rownames(dds.adult.grp.DESeq) %in% rownames(diffex.adult.FB)) 
diffex.adult.FB.counts <- diffex.adult.FB.counts[,subset(key.filtered, stage=="adult" & population=="Fidalgo Bay")$sample]

diffex.adult.DB.counts <- subset(counts(dds.adult.grp.DESeq), rownames(dds.adult.grp.DESeq) %in% rownames(diffex.adult.DB))
diffex.adult.DB.counts <- diffex.adult.DB.counts[,subset(key.filtered, stage=="adult" & population=="Dabob Bay")$sample]

diffex.adult.OB1.counts <- subset(counts(dds.adult.grp.DESeq), rownames(dds.adult.grp.DESeq) %in% rownames(diffex.adult.OB1))
diffex.adult.OB1.counts <- diffex.adult.OB1.counts[,subset(key.filtered, stage=="adult" & population=="Oyster Bay C1")$sample]
```

# generate heatmap with differentially expressed genes within FB 

```{r}
dds.adult.df.FB <- as.data.frame(colData(dds.adult)[colData(dds.adult)$population=="Fidalgo Bay",c("population", "pCO2.parent")])
dds.adult.df.FB <- dds.adult.df.FB[match(colnames(diffex.adult.FB.counts), rownames(dds.adult.df.FB)),]
all(colnames(diffex.adult.FB.counts) == rownames(dds.adult.df.FB)) #double check that samples are still in same order 

pheatmap(diffex.adult.FB.counts, cluster_rows=T, show_rownames=FALSE, cluster_columns=T, na.rm=TRUE, 
         scale="row", main = "(Adults) FB DEGs by direct pCO2 exposure, clustered", 
         annotation_col=dds.adult.df.FB[2], color=c("dodgerblue3", "goldenrod1"))
```

# generate heatmap with differentially expressed genes within DB

```{r}
dds.adult.df.DB <- as.data.frame(colData(dds.adult)[colData(dds.adult)$population=="Dabob Bay",c("population", "pCO2.parent")])
dds.adult.df.DB <- dds.adult.df.DB[match(colnames(diffex.adult.DB.counts), rownames(dds.adult.df.DB)),]
all(colnames(diffex.adult.DB.counts) == rownames(dds.adult.df.DB)) #double check that samples are still in same order 

pheatmap(diffex.adult.DB.counts, cluster_rows=T, show_rownames=FALSE, cluster_columns=T, na.rm=TRUE, scale="row", 
         main = "(Adults) DB DEGs by direct pCO2 exposure, clustered", annotation_col=dds.adult.df.DB[2], 
         color=c("dodgerblue3", "goldenrod1"))
```

```{r}
adult.FB.p05.names <-rownames(diffex.adult.FB[order(diffex.adult.FB$padj),])
adult.DB.p05.names <-rownames(diffex.adult.DB[order(diffex.adult.DB$padj),])
count_plots = list()

# #plot FB DEGs to find any 
# for (i in 1:length(adult.FB.p05.names)) {
# a <-  plotCounts(dds.adult.DESeq, gene=adult.FB.p05.names[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
# b <- ggplot(subset(a, population=="Fidalgo Bay") %>% rownames_to_column("sample"),
#        aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
#   geom_point(position=position_jitter(w = 0.1,h = 0)) +
#     geom_text() +
#     theme_bw() +
#     ggtitle(adult.FB.p05.names[i]) +
#     theme(plot.title = element_text(hjust = 0.5))
# count_plots[[i]] <- b
# }
# count_plots
# 
# #plot DB DEGs to find any issues 
# for (i in 1:length(adult.DB.p05.names)) {
# a <-  plotCounts(dds.adult.DESeq, gene=adult.DB.p05.names[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
# b <- ggplot(subset(a, population=="Dabob Bay") %>% rownames_to_column("sample"),
#        aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
#   geom_point(position=position_jitter(w = 0.1,h = 0)) +
#     geom_text() +
#     theme_bw() +
#     ggtitle(adult.DB.p05.names[i]) +
#     theme(plot.title = element_text(hjust = 0.5))
# count_plots[[i]] <- b
# }
# count_plots
```

## Genes that aren't convincing (aka only 1 or 2 samples had higher expression in a treatment): 

### Fidalgo Bay
- OLUR_00026074  
- OLUR_00015132  
- OLUR_00018477    
- OLUR_00024224  

### Dabob Bay 
- OLUR_00004613  
- OLUR_00013617  
- OLUR_00021421  
- OLUR_00020678
- OLUR_00031169  
- OLUR_00016878  
- OLUR_00000822  
- OLUR_00003610  
- OLUR_00000097  
- OLUR_00001769  
- OLUR_00019562  
- OLUR_00004194  

# Plot single differentially expressed genes within OB1

```{r}
diffex.adult.OB1.counts <- subset(counts(dds.adult.grp.DESeq), rownames(dds.adult.grp.DESeq) %in% rownames(diffex.adult.OB1))
#diffex.adult.OB1.counts <- diffex.adult.OB1.counts[,subset(key.filtered, stage=="adult" & population=="Oyster Bay C1")$sample]

a <-  plotCounts(dds.adult.DESeq, gene=rownames(diffex.adult.OB1.counts), intgroup=c("pCO2.parent"), returnData = TRUE)
ggplotly(ggplot(subset(a, population!="Oyster Bay C1") %>% rownames_to_column("sample") %>% drop_na(),
       aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
  geom_point(size=3, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text(size=3) +
    theme_bw() +
    ggtitle(rownames(diffex.adult.OB1.counts)) +
    theme(plot.title = element_text(hjust = 0.5, size = 12), axis.text.x = element_blank(),
          axis.text.y = element_text(size=6),
          axis.title = element_blank(), axis.ticks.x = element_blank(), legend.position = "none"))
```

# Heat map of any pCO2-related DEG (all pops, DB, and FB) combined, showing all 3 populations. 

```{r}
diffex.adult.counts <- subset(counts(dds.adult.DESeq), rownames(dds.adult.DESeq) %in% unique(c(rownames(diffex.adult), rownames(diffex.adult.DB), rownames(diffex.adult.FB))))
diffex.adult.counts <- diffex.adult.counts[,subset(key.filtered, stage=="adult")$sample]
dds.adult.df <- as.data.frame(colData(dds.adult)[,c("population", "pCO2.parent")])
dds.adult.df <- dds.adult.df[match(colnames(diffex.adult.counts), rownames(dds.adult.df)),]
all(colnames(diffex.adult.counts) == rownames(dds.adult.df)) #double check that samples are still in same order 

pheatmap(diffex.adult.counts[,rownames(dds.adult.df %>% filter(population=="Dabob Bay"))], cluster_rows=T, cluster_cols = F,
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, 
         scale="row", main = "(Adults) All DEGs by direct pCO2 exposure, Dabob Bay", 
         annotation_col=dds.adult.df[2])

pheatmap(diffex.adult.counts[,rownames(dds.adult.df %>% filter(population=="Fidalgo Bay"))], cluster_rows=T, cluster_cols = F,
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, 
         scale="row", main = "(Adults) All DEGs by direct pCO2 exposure, Fidalgo Bay", 
         annotation_col=dds.adult.df[2])

pheatmap(diffex.adult.counts[,rownames(dds.adult.df %>% filter(population=="Oyster Bay C1"))], cluster_rows=T, cluster_cols = F,
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, 
         scale="row", main = "(Adults) All DEGs by direct pCO2 exposure, Oyster Bay", 
         annotation_col=dds.adult.df[2])
```


## Identify functions of DEGs 

# IMPORTANT: the dataframe "adult.all.DEGs" contains all DEGs for each population with annotation information. These DEGs were identified for each population separately. 

```{r}
(adult.all.DEGs <- left_join(
  rbind(as.data.frame(diffex.adult.FB) %>%
  mutate(population="FB"),
  as.data.frame(diffex.adult.DB) %>%
  mutate(population="DB"),
  as.data.frame(diffex.adult.OB1) %>%
  mutate(population="OB1")) %>%
  rownames_to_column(var = "Name"),
  Olurida_gene_uniprot %>% dplyr::select(contig, start, end, Name, Note, SPID),
  by="Name")) 
```

Any DEGs the same in multiple populations?

```{r}
Reduce(intersect, list(rownames(diffex.adult.FB),rownames(diffex.adult.DB), rownames(diffex.adult.OB1))) #none 
Reduce(intersect, list(rownames(diffex.adult.FB),rownames(diffex.adult.DB))) #4 the same 
Reduce(intersect, list(rownames(diffex.adult.FB),rownames(diffex.adult.OB1))) #none
Reduce(intersect, list(rownames(diffex.adult.DB),rownames(diffex.adult.OB1))) #none

#Any DEGs in the all-pop analysis also in within-pop analyses?
Reduce(intersect, list(rownames(diffex.adult),rownames(diffex.adult.FB))) %>% length()  #yes, 10
Reduce(intersect, list(rownames(diffex.adult),rownames(diffex.adult.DB))) %>% length() #yes, 32
Reduce(intersect, list(rownames(diffex.adult),rownames(diffex.adult.OB1))) %>% length() #no
Reduce(intersect, list(rownames(diffex.adult),rownames(diffex.adult.FB), rownames(diffex.adult.DB))) %>% length()  #the 4 common in FB & DB are also found in the all-pop model 
```

What are the genes that are DEGs in both FB and DB?

```{r}
(adult.all.DEGs.common <- adult.all.DEGs %>% 
   filter(Name %in% Reduce(intersect, list(rownames(diffex.adult.FB),rownames(diffex.adult.DB)))))
```

Plot counts for the 4 DEGs common in both FB and DB 

```{r}
adult.all.DEGs.names <-adult.all.DEGs.common$Name

adult.all.DEGs.common <- adult.all.DEGs.common %>% dplyr::select(Name, SPID) %>% 
  mutate(Protein=c("Cytochrome P450 2B4 (CYP2B4)", 
                   "Fatty acid-binding protein, adipocyte (FABP4)", 
                   "Cytochrome P450 2U1 (CYP2UI)", "Thyroxine 5-deiodinase (DIO3)" ))

count_plots.adult = list()
for (i in 1:length(adult.all.DEGs.names)) {
a <-  plotCounts(dds.adult.DESeq, gene=adult.all.DEGs.names[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
b <- ggplot(a %>% rownames_to_column("sample"), #subset(a, population!="Oyster Bay C1")
       aes(x=population:pCO2.parent, y=count, color=population, shape=pCO2.parent, label=sample)) +
  geom_point(size=3, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text(size=3) +
    theme_bw() +
    ggtitle(adult.all.DEGs.common[i, "Protein"]) +
    theme(plot.title = element_text(hjust = 0.5, size = 10), axis.text.x = element_blank(),
          axis.text.y = element_text(size=6),
          axis.title = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
  scale_color_manual(values = c("#4daf4a","#377eb8", "#984ea3"), name="Cohort", guide = guide_legend(override.aes = list(size=4))) +
  scale_shape(name="pCO2 exposure", guide = guide_legend(override.aes = list(size=4))) 
count_plots.adult[[i]] <- b
}
plot_grid(plotlist=count_plots.adult) #plot them in a grid 
#count_plots.juv.DB #plot them all individually 
```

Enrichment Analysis! 

```{r}
# DEG in FB 
diffex.adult.FB %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID, Note), by = c("gene"="Name")) %>% 
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# All genes in FB (background)
res.adult.FB %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID), by = c("gene"="Name")) %>% 
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEG in DB 
diffex.adult.DB %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot, by = c("gene"="Name")) %>% View() # %>% dplyr::select(Name, SPID)
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# All genes in DB (background)
res.adult.DB %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID), by = c("gene"="Name")) %>% 
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Fidalgo Bay - Enriched functions in DEGs between pCO2 exposure 

| Term | Count | % | PValue | Genes | List Total | Pop Hits | Pop Total | Fold Enrichment | FDR |
|-|-|-|-|-|-|-|-|-|-|
| GO:0030097~hemopoiesis | 3 | 7.5 | 0.0126 | Q04721, Q02858, P08953 | 38 | 39 | 8403 | 17.01 | 1.0 |
| GO:1990403~embryonic brain development | 2 | 5.0 | 0.0175 | E9PY46, P62997 | 38 | 4 | 8403 | 110.57 | 1.0 |
| GO:0007507~heart development | 4 | 10.0 | 0.0208 | P97864, E9PY46, Q02858, P08953 | 38 | 134 | 8403 | 6.60 | 1.0 |
| GO:0044255~cellular lipid metabolic process | 2 | 5.0 | 0.0516 | Q9BYK8, Q08BB2 | 38 | 12 | 8403 | 36.86 | 1.0 |
| GO:0009617~response to bacterium | 2 | 5.0 | 0.0723 | P08953, Q6AYS4 | 38 | 17 | 8403 | 26.015 | 1.0 |
| GO:0045944~positive regulation of transcription from RNA polymerase II promoter | 5 | 12.5 | 0.0868 | Q7ZVK3, Q62725, Q9BYK8, E9QAM5, P08953 | 38 | 384 | 8403 | 2.88 | 1.0 |

## Dabob Bay - Enriched functions in DEGs between pCO2 exposure 

| Term | Count | % | PValue | Genes | List Total | Pop Hits | Pop Total | Fold Enrichment | FDR |
|-|-|-|-|-|-|-|-|-|-|
| GO:0055114~oxidation-reduction process | 12 | 17.65 | 1.04E-4 | Q9Z339, P49894, P11714, P51174, Q64583, P10632, O42412, P34277, Q7Z449, Q0IIF9, Q6VVW9, Q9R0H0 | 57 | 431 | 8403 | 4.1 | 0.023 |
| GO:0002591~positive regulation of antigen processing and presentation of peptide antigen via MHC class I | 3 | 4.41 | 2.60E-4 | P08183, P21440, P06795 | 57 | 4 | 8403 | 110.6 | 0.023 |
| GO:0002489~antigen processing and presentation of endogenous peptide antigen via MHC class Ib via ER pathway, TAP-dependent | 3 | 4.41 | 2.60E-4 | P08183, P21440, P06795 | 57 | 4 | 8403 | 110.6 | 0.023 |
| GO:0002485~antigen processing and presentation of endogenous peptide antigen via MHC class I via ER pathway, TAP-dependent | 3 | 4.41 | 2.60E-4 | P08183, P21440, P06795 | 57 | 4 | 8403 | 110.6 | 0.023 |
| GO:0002481~antigen processing and presentation of exogenous protein antigen via MHC class Ib, TAP-dependent | 3 | 4.41 | 2.60E-4 | P08183, P21440, P06795 | 57 | 4 | 8403 | 110.6 | 0.023 |
| GO:0006631~fatty acid metabolic process | 4 | 5.88 | 0.0049 | P48035, O35488, P51174, Q9R0H0 | 57 | 52 | 8403 | 11.3 | 0.36 |
| GO:0008152~metabolic process | 6 | 8.82 | 0.0071 | Q9Z339, O35488, P51174, P34277, P47713, Q9R0H0 | 57 | 182 | 8403 | 4.9 | 0.45 |
| GO:0055088~lipid homeostasis | 3 | 4.41 | 0.0084 | P51174, P21440, Q9R0H0 | 57 | 21 | 8403 | 21.1 | 0.47 |
| GO:0006805~xenobiotic metabolic process | 3 | 4.41 | 0.010 | Q95JD5, P10632, O43704 | 57 | 23 | 8403 | 19.2 | 0.50 |
| GO:0006810~transport | 9 | 13.23 | 0.011 | Q9Z2J0, P48035, P08183, Q8R4G9, Q7TM99, P21440, P06795, Q99758, P43006 | 57 | 466 | 8403 | 2.9 | 0.51 |
| GO:0042493~response to drug | 5 | 7.35 | 0.013 | P08183, Q8R4G9, P06795, Q99758, P43006 | 57 | 137 | 8403 | 5.4 | 0.52 |
| GO:0055085~transmembrane transport | 6 | 8.82 | 0.015 | Q9Z2J0, P08183, Q7TM99, P21440, P06795, Q99758 | 57 | 221 | 8403 | 4.0 | 0.57 |
| GO:0097267~omega-hydroxylase P450 pathway | 2 | 2.94 | 0.020 | P10632, Q7Z449 | 57 | 3 | 8403 | 98.3 | 0.68 |
| GO:0030855~epithelial cell differentiation | 3 | 4.41 | 0.027 | Q9UGM3, Q95JD5, O43704 | 57 | 39 | 8403 | 11.3 | 0.87 |
| GO:0008202~steroid metabolic process | 3 | 4.41 | 0.037 | Q95JD5, P10632, O43704 | 57 | 46 | 8403 | 9.6 | 1 |
| GO:0006940~regulation of smooth muscle contraction | 2 | 2.94 | 0.039 | P49817, Q8R4G9 | 57 | 6 | 8403 | 49.1 | 1 |
| GO:0009812~flavonoid metabolic process | 2 | 2.94 | 0.052 | Q95JD5, O43704 | 57 | 8 | 8403 | 36.9 | 1 |
| GO:0006629~lipid metabolic process | 4 | 5.88 | 0.054 | O35488, P51174, P47713, Q9R0H0 | 57 | 129 | 8403 | 4.6 | 1 |
| GO:0042446~hormone biosynthetic process | 2 | 2.94 | 0.065 | P49894, O42412 | 57 | 10 | 8403 | 29.5 | 1 |
| GO:0045332~phospholipid translocation | 2 | 2.94 | 0.065 | P08183, P21440 | 57 | 10 | 8403 | 29.5 | 1 |
| GO:0042403~thyroid hormone metabolic process | 2 | 2.94 | 0.071 | Q95JD5, O43704 | 57 | 11 | 8403 | 26.8 | 1 |
| GO:0006855~drug transmembrane transport | 2 | 2.94 | 0.083 | P08183, P21440 | 57 | 13 | 8403 | 22.7 | 1 |
| GO:0010942~positive regulation of cell death | 2 | 2.94 | 0.089 | O95069, Q8CJ26 | 57 | 14 | 8403 | 21.1 | 1 |
| GO:0000038~very long-chain fatty acid metabolic process | 2 | 2.94 | 0.096 | O35488, Q9R0H0 | 57 | 15 | 8403 | 19.7 | 1 |
| GO:0051923~sulfation | 2 | 2.94 | 0.096 | Q95JD5, O43704 | 57 | 15 | 8403 | 19.7 | 1 |


Read in GO terms for Revigo

```{r}
read_delim(file="../results/Adult_FB_DEGs_EnrichedBP.txt", delim = "\t") %>% 
  mutate(GO = str_extract(Term, "GO(.*?)~")) %>% 
  mutate(GO = gsub("~", "", GO)) %>% dplyr::select(GO, PValue) %>% na.omit() %>% write_clip()

read_delim(file="../results/Adult_DB_DEGs_EnrichedBP.txt", delim = "\t") %>% 
  mutate(GO = str_extract(Term, "GO(.*?)~")) %>% 
  mutate(GO = gsub("~", "", GO)) %>% dplyr::select(GO, PValue) %>% na.omit() %>% write_clip()
```
Script to extract GO terms for all DEGs 

```{r}
read.csv(file="../results/Adult_FB_DEGs.csv") %>% 
  mutate(GO = strsplit(as.character(Gene.ontology.IDs), ";")) %>% 
    unnest(GO) %>%
dplyr::select(Entry, GO) %>% na.omit() #%>% unlist() %>% as.vector() %>% write_clip()
```

```{r}
## Add a factor that defines interaction between pop and pCO2 so I can single out ambient  
dds.adult.DESeq$pop.pco2 <- factor(paste0(dds.adult.DESeq$population, dds.adult.DESeq$pCO2.parent))
design(dds.adult.DESeq) <- ~ pop.pco2
dds.adult.DESeq <- DESeq(dds.adult.DESeq, betaPrior=TRUE)
```


```{r}
# Here are all the possible contrasts I can make 
levels(dds.adult.DESeq$pop.pco2)

print("Comparison: Dabob Bay vs. Fidalgo & Oyster Bay")
summary(res.adult.DBvs <- results(dds.adult.DESeq, contrast=list(c("pop.pco2Dabob.BayAmbient"), c("pop.pco2Fidalgo.BayAmbient","pop.pco2Oyster.Bay.C1Ambient")), listValues=c(1, -1/2)))

print("Comparison: Fidalgo Bay vs. Dabob & Oyster Bay C1")
summary(res.adult.FBvs <- results(dds.adult.DESeq, contrast=list(c("pop.pco2Fidalgo.BayAmbient"), c("pop.pco2Dabob.BayAmbient","pop.pco2Oyster.Bay.C1Ambient")), listValues=c(1, -1/2)))

print("Comparison: Oyster Bay C1 vs. Dabob & Fidalgo")
summary(res.adult.OB1vs <- results(dds.adult.DESeq, contrast=list(c("pop.pco2Oyster.Bay.C1Ambient"), c("pop.pco2Fidalgo.BayAmbient","pop.pco2Dabob.BayAmbient")), listValues=c(1, -1/2)))
```
## Count # of genes diff expressed  (p-value <0.05) in between each population and the others 

```{r}
paste("No. of genes differentially expressed (padj<0.05) between Dabob & other cohorts, adults:",  sum(res.adult.DBvs$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between Fidalgo & other cohorts, adults:",  sum(res.adult.FBvs$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between Oyster Bay C1 & other cohorts, adults:",  sum(res.adult.OB1vs$padj < 0.05, na.rm=TRUE))
```

## Save population vs. population DEG lists for later comparisons 

```{r}
diffex.adult.DBvs <- subset(res.adult.DBvs, padj < 0.05)
diffex.adult.FBvs <- subset(res.adult.FBvs, padj < 0.05)
diffex.adult.OB1vs <- subset(res.adult.OB1vs, padj < 0.05)

# save all R objects to file for integration 
save(diffex.adult.DBvs, file="../results/diffex.adult.DBvs")
save(diffex.adult.FBvs, file="../results/diffex.adult.FBvs")
save(diffex.adult.OB1vs, file="../results/diffex.adult.OB1vs")

# merge and select unique genes for PCA 
degs.adult.pops <- c(rownames(diffex.adult.DBvs), 
  rownames(diffex.adult.FBvs),
  rownames(diffex.adult.OB1vs)) %>% unique()
save(degs.adult.pops, file="../results/degs.adult.pops")
```

```{r}
# PCA using genes that were DEGs in 1 population vs other 2 contrast 
ggplotly(
  plotPCA(vsd.adult[degs.adult.pops,adult.amb], intgroup="population") + 
           ggtitle("PCA by cohorts, ambient only, DEGs among cohorts") + 
    geom_point(size=3, aes(text=colnames(vsd.adult[degs.adult.pops,adult.amb]))) + 
    theme_minimal(), tooltip = "text")
```

```{r}
# generate heatmap of population DEGs  
vsd.adult.df <- as.data.frame(colData(vsd.adult)[,c("population", "pCO2.parent")])
adult.amb <- (key.filtered %>% filter(stage=="adult" & pCO2.parent=="Ambient"))$sample

out <- pheatmap(assay(vsd.adult[degs.adult.pops,adult.amb]), cluster_rows=T, show_rownames=FALSE, show_colnames=TRUE,
         cluster_cols=F, annotation_col=vsd.adult.df[1], scale="row", main = "DEGs among cohorts, Adults",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
```


```{r}
# extract gene clusters 
hc <-out$tree_row
lbl <- cutree(hc, 9) # split gene dendrogram in 4 groups
which(lbl==1) %>% names() # grab genes of first group
```

# Extract counts for the gene clusters 
```{r}
clust1 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==1) %>% names(),]
clust2 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==2) %>% names(),]
clust3 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==3) %>% names(),]
clust4 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==4) %>% names(),]
clust5 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==5) %>% names(),]
clust6 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==6) %>% names(),]
clust7 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==7) %>% names(),]
clust8 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==8) %>% names(),]
clust9 <- assay(vsd.adult[degs.adult.pops,adult.amb])[which(lbl==9) %>% names(),]

pheatmap(clust3, cluster_rows=T, show_rownames=FALSE, show_colnames=TRUE,
         cluster_cols=F, annotation_col=vsd.adult.df[1], scale="row", main = "DEGs among cohorts, Adults",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
# cluster 3 is the problem child - remove those genes from the DBvs gene set prior to enrichment analysis 

rownames(clust3) #here's how I pull those gene names 

# Here's the heat map with those troublesome genes removed. Looks way better! 
pheatmap(assay(vsd.adult[degs.adult.pops[!degs.adult.pops %in% rownames(clust3)], adult.amb]), cluster_rows=T, show_rownames=FALSE, show_colnames=TRUE,
         cluster_cols=F, annotation_col=vsd.adult.df[1], scale="row", main = "DEGs among cohorts, Adults",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
```


Enrichment Analysis using DEGs identified 1 population vs other pops comparisons (ambient pCO2 only)

```{r}
# DEGs between DB & other cohorts 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.adult.DBvs)) %>% filter(!Name %in% rownames(clust3)) %>%
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between DB & OB1
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.adult.FBvs)) %>% filter(!Name %in% rownames(clust3)) %>%
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between FB & OB1
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.adult.OB1vs.amb)) %>% filter(!Name %in% rownames(clust3)) %>%
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# All genes for background
filter(Olurida_gene_uniprot, 
       Name %in% c(rownames(res.adult.DBvs) %>% unique())) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

# Here's another option for population expression differences - Examine DEGs among population in pairwise comparisons <- THIS IS WHAT I USE IN THE END 

```{r}
key.filtered %>% filter(stage=="adult" & pCO2.parent=="Ambient") %>% group_by(population, pCO2.parent) %>% tally() # 9 per pop 
keep2 <- (key.filtered %>% filter(stage=="adult" & pCO2.parent=="Ambient"))$sample
dds.adult.ambient <- dds.adult[,keep2]
design(dds.adult.ambient) <- formula(~ population)
dds.adult.ambient.DESeq <- DESeq(dds.adult.ambient)
```
## Visualize sample clustering via PCA (after transformation)

```{r}
vsd.adult.ambient <- varianceStabilizingTransformation(dds.adult.ambient, blind=FALSE)

#heatmap with all genes - save to file since it's so large 
# pheatmap(log2(assay(vsd.adult.ambient)), cluster_rows=T, show_rownames=FALSE,
#          cluster_cols=F, annotation_col=as.data.frame(colData(vsd.adult.ambient)[1]),
#          filename = "../results/population-comparisons-july2021/adult-heatmap-allgenes.png")

# PCA with points color coded by population 
ggplotly(
  plotPCA(vsd.adult.ambient, intgroup="population") + 
           ggtitle("PCA by population, ambient only (var-stabilizing transformed)") + 
    geom_point(size=3, aes(text=colnames(vsd.adult.ambient))) + 
    theme_minimal()+ stat_ellipse(), tooltip = "text")
```

```{r}
print("Comparison: cohort, ambient only")
summary(res.adult.DBFB.amb <- results(dds.adult.ambient.DESeq, contrast=c("population", "Dabob Bay", "Fidalgo Bay"), alpha=0.05))
summary(res.adult.DBOB1.amb <- results(dds.adult.ambient.DESeq, contrast=c("population", "Dabob Bay", "Oyster Bay C1"), alpha=0.05))
summary(res.adult.FBOB1.amb <- results(dds.adult.ambient.DESeq, contrast=c("population", "Fidalgo Bay", "Oyster Bay C1"), alpha=0.05))

paste("No. of genes differentially expressed (padj<0.05) between DB & FB, ambient adults only:",  sum(res.adult.DBFB.amb$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between DB & OB1, ambient adults only:",  sum(res.adult.DBOB1.amb$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between FB & OB1, ambient adults only:",  sum(res.adult.FBOB1.amb$padj < 0.05, na.rm=TRUE))

#Volcano plots of ambient pairwise 
plotMA(res.adult.DBFB.amb, ylim=c(-6,6), main="log2 FC / mean normalized counts (ambient co2 only)\nDEGs between DB & FB")
plotMA(res.adult.DBOB1.amb, ylim=c(-6.5,7.25), main="log2 FC / mean normalized counts (ambient co2 only)\nDEGs between DB & OB")
plotMA(res.adult.FBOB1.amb, ylim=c(-6.5,5.5), main="log2 FC / mean normalized counts (ambient co2 only)\nDEGs between FB & OB")
```

## Save population vs. population DEG lists

```{r}
diffex.adult.DBFB.amb <- subset(res.adult.DBFB.amb, padj < 0.05)
diffex.adult.DBOB1.amb <- subset(res.adult.DBOB1.amb, padj < 0.05)
diffex.adult.FBOB1.amb <- subset(res.adult.FBOB1.amb, padj < 0.05)

# save all R objects to file for integration 
save(diffex.adult.DBFB.amb, file="../results/diffex.adult.DBFB.amb")
save(diffex.adult.DBOB1.amb, file="../results/diffex.adult.DBOB1.amb")
save(diffex.adult.FBOB1.amb, file="../results/diffex.adult.FBOB1.amb")

# merge and select unique genes that are DEGs among any pairwise population comparison  
degs.adult.pops.amb <- c(rownames(diffex.adult.DBFB.amb), 
  rownames(diffex.adult.DBOB1.amb),
  rownames(diffex.adult.FBOB1.amb)) %>% unique()
save(degs.adult.pops.amb, file="../results/degs.adult.pops.amb")
```

```{r}
# generate heatmap of population DEGs  
vsd.adult.ambient.df <- as.data.frame(colData(vsd.adult.ambient)[,c("population", "pCO2.parent")]) %>% mutate(Cohort=population)

print(out.adult <- pheatmap(assay(vsd.adult.ambient[degs.adult.pops.amb,]), cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=T, annotation_col=vsd.adult.ambient.df[3], scale="row", main = "DEGs among cohorts, Adults (ambient pCO2 only)",
         annotation_colors=list(Cohort=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")), 
         cutree_cols = 3))
```

```{r}
# extract gene clusters 
hc.adult <-out.adult$tree_row

# identify which "height" to cut tree 
plot(hc.adult)
abline(h=8.45, col="red", lty=2, lwd=2) #9.25 = 4 clusters, 8.5 = 7 clusters

lbl.adult <- cutree(hc.adult, h=8.45) # split gene dendrogram into 5 clusters, alternatively could use "k=12" 
#which(lbl==1) %>% names() # grab genes of first group

# Extract gene names for the  gene clusters and plot
cluster1.adult <- assay(vsd.adult.ambient)[which(lbl.adult==1) %>% names(),]
cluster2.adult <- assay(vsd.adult.ambient)[which(lbl.adult==2) %>% names(),]
cluster3.adult <- assay(vsd.adult.ambient)[which(lbl.adult==3) %>% names(),]
cluster4.adult <- assay(vsd.adult.ambient)[which(lbl.adult==4) %>% names(),]
cluster5.adult <- assay(vsd.adult.ambient)[which(lbl.adult==5) %>% names(),]
cluster6.adult <- assay(vsd.adult.ambient)[which(lbl.adult==6) %>% names(),]
cluster7.adult <- assay(vsd.adult.ambient)[which(lbl.adult==7) %>% names(),]

# Plot each cluster separately to inspect pattern 
pheatmap(cluster1.adult, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row", main = "Adult cluster 1 = HHL ",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
pheatmap(cluster2.adult, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row",  main = "Adult cluster 2 = HML",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
pheatmap(cluster3.adult, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row",  main = "Adult cluster 3 = LLH",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
pheatmap(cluster4.adult, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row",  main = "Adult cluster 4 = LHH",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
pheatmap(cluster5.adult, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row",  main = "Adult cluster 5 = HLL",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
pheatmap(cluster6.adult, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row", main = "Adult cluster 6 = LHL",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
pheatmap(cluster7.adult, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row",  main = "Adult cluster 7 = HLL (but only 2 outliers in DB)",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
```

High in DB: 2,5
Low in DB: 4
High in FB: 6
Low in FB: none 
High in OB: 3
Low in OB: 1

# Enrichment analysis on each gene cluster 

```{r}
# Background list 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(assay(vsd.adult.ambient))) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Genes HIGHER in DB - clusters 2 & 5
filter(Olurida_gene_uniprot, 
       Name %in% c(rownames(cluster2.adult), rownames(cluster5.adult))) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Genes LOWER in DB - cluster 4
filter(Olurida_gene_uniprot, 
       Name %in% rownames(cluster4.adult)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Genes HIGHER in FB - clusters 6
filter(Olurida_gene_uniprot, 
       Name %in% rownames(cluster6.adult)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Genes LOWER in FB - none

# Genes HIGHER in OB - cluster 3
filter(Olurida_gene_uniprot, 
       Name %in% rownames(cluster3.adult)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Genes LOWER in OB - cluster 1
filter(Olurida_gene_uniprot, 
       Name %in% rownames(cluster1.adult)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Saved into the results/population-comparisons-july2021/ folder 
```


## Are genes constitutively expressed at different levels among populations (pairwise-population DEGs) also diff. expressed in high pCO2? 

```{r}
# Dabob Bay
DB.a<- Reduce(intersect, list(rownames(diffex.adult.DBFB.amb), rownames(diffex.adult.DB))) # Genes diff. expressed in DB in pCO2, & also between DB & FB at ambient
DB.b<- Reduce(intersect, list(rownames(diffex.adult.DBOB1.amb), rownames(diffex.adult.DB))) # Genes diff. expressed in DB in pCO2, & also between DB & OB1 at ambient
DB.c<- Reduce(intersect, list(rownames(diffex.adult.DBvs)[!rownames(diffex.adult.DBvs) %in% rownames(clust3)]), rownames(diffex.adult.DB)) # Genes diff. expressed in DB in pCO2, & also between DB & two other 

# Here are annotated genes that were constitutively expressed at diff levels in DB compared to FB or OB!, but also diff express in high pCO2 
filter(Olurida_gene_uniprot, 
       Name %in% (c(DB.a,DB.b,DB.c) %>% unique())) %>% filter(!is.na(SPID))

# Fidalgo Bay 
FB.a<- Reduce(intersect, list(rownames(diffex.adult.DBFB.amb), rownames(diffex.adult.FB))) # Genes diff. expressed in DB in pCO2, & also between DB & FB at ambient
FB.b<- Reduce(intersect, list(rownames(diffex.adult.FBOB1.amb), rownames(diffex.adult.FB))) # Genes diff. expressed in DB in pCO2, & also between DB & OB1 at ambient
FB.c<- Reduce(intersect, list(rownames(diffex.adult.FBvs), rownames(diffex.adult.FB))) # Genes diff. expressed in DB in pCO2, & also between DB & two other 

# Here are annotated genes that were constitutively expressed at diff levels in DB compared to FB or OB!, but also diff express in high pCO2 
filter(Olurida_gene_uniprot, 
       Name %in% (c(FB.a,FB.b, FB.c) %>% unique())) %>% filter(!is.na(SPID))

```


Enrichment Analysis using DEGs identified in  pairwise population comparisons (ambient pCO2 only)

```{r}
# DEGs between DB & FB
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.adult.DBFB.amb)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between DB & OB1
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.adult.DBOB1.amb)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between FB & OB1
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.adult.FBOB1.amb)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# All genes for background
filter(Olurida_gene_uniprot, 
       Name %in% c(rownames(res.adult.DBFB.amb),rownames(res.adult.DBOB1.amb),rownames(res.adult.FBOB1.amb) %>% unique())) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

# RESULTS SYNOPSIS FOR PAPER - ADULTS 

```{r}
# SUMMARY STATS OF FILTERED READS/GENES READY FOR ANSLYSES 
print(paste("# of genes dropped before analysis:", ncol(counts.t.adult) - ncol(counts.ts.adult), sep=" "))
print(paste("Total # genes for analysis:", ncol(counts.ts.adult)))

a <- data.frame(rowSums(counts.t.adult != 0)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample") %>% summarise(mean=mean(count.total), sd=sd(count.total), se=sd/sqrt(length(sample)))
print(paste("Per-sample average no. of genes:", round(a$mean), "+/- SD", round(a$sd)))

b <- data.frame(rowSums(counts.t.adult)) %>% 
                  dplyr::rename(read.total = 1) %>% 
                  rownames_to_column(var="sample") %>% 
summarise(min=min(read.total), max=max(read.total), mean=mean(read.total), sd=sd(read.total), se=sd/sqrt(length(sample))) 
print(paste("Per-sample range and average no. of reads with error rates: ", round(b$min), "-", round(b$max), " ", round(b$mean), " +/- SD ", round(b$sd), sep=""))

# DEG ANAYSIS
paste("No. of genes differentially expressed (padj<0.05) by pCO2, adult ctendia, comparison across all pops:",  sum(res.adult.pco2$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Fidalgo Bay adult ctenidia:",  sum(res.adult.FB$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Dabob Bay adult ctenidia:",  sum(res.adult.DB$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Oyster Bay Cohort1 adult ctenidia:",  sum(res.adult.OB1$padj < 0.05, na.rm=TRUE))

#check out how many are more abundant and less abundant in high pCO2, and % of all genes that are DEGs  
summary(res.adult.pco2) 
summary(res.adult.FB)
summary(res.adult.DB)

#Enriched Biological Processes 
EnrBP.adult.pCO2.allpops #In response to pCO2 treatment, all-populations combined, adults 
EnrBP.adult.pCO2.FB #In response to pCO2 treatment, Fidalgo Bay, adults 
EnrBP.adult.pCO2.DB #In response to pCO2 treatment, Dabob Bay combined, adults

# Are any GO terms common amon all pops, FB and DB? 
Reduce(intersect, list(EnrBP.adult.pCO2.allpops$go,EnrBP.adult.pCO2.FB$go, EnrBP.adult.pCO2.DB$go)) #none 
```

# Bubble plots for paper 

Need dataframe with these columns: 
- (x) group= c(all.pops, FB, DB, OB)  
- (y) Term= enriched GO term 
- (alpha) p.value.adjusted = enrichment analysis FDR adjusted p-values  
- (color) high.transcripts = c(Greater, Fewer), aka the number of transcripts in high pCO2 compaerd to ambient 
- (facet_grid) GO.category (biological processes vs. molecular function)
- (labeller) GOterm_names 

```{r}
# left_join(GOslim, by=c("go"="GO_ID"))

EnrBP.adult.pCO2.allpops <- read_delim(file="../results/Adult_allpops_DEGs_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(group="All Cohorts", GO.category="Biological Processes")

EnrMF.adult.pCO2.allpops <- read_delim(file="../results/Adult_allpops_DEGs_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(group="All Cohorts", GO.category="Molecular Function")

EnrBP.adult.pCO2.FB <- read_delim(file="../results/Adult_FB_DEGs_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(group="Fidalgo Bay", GO.category="Biological Processes")

EnrMF.adult.pCO2.FB <- read_delim(file="../results/Adult_FB_DEGs_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(group="Fidalgo Bay", GO.category="Molecular Function")

EnrBP.adult.pCO2.DB <- read_delim(file="../results/Adult_DB_DEGs_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(group="Dabob Bay", GO.category="Biological Processes")

EnrMF.adult.pCO2.DB <- read_delim(file="../results/Adult_DB_DEGs_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(group="Dabob Bay", GO.category="Molecular Function")

all.go.enriched <- rbind(
  EnrBP.adult.pCO2.allpops,
  EnrMF.adult.pCO2.allpops,
  EnrBP.adult.pCO2.FB,
  EnrMF.adult.pCO2.FB,
  EnrBP.adult.pCO2.DB,
  EnrMF.adult.pCO2.DB)

# Here I write out the merged enriched GO term dataframe so I can add a column by hand ("Abundance.HighPCO2") which indicates whether that process/function was more or less abundant in the High pCO2 treatment. 
write.csv(all.go.enriched %>% left_join(GOslim, by=c("go"="GO_ID")), file="../results/Adults.go.enriched.4paper.csv")  
```

# For each GO term, assess whether counts were MORE abundant or LESS abundant in the high pCO2 exposed adults 

```{r}
# Here I plot all the genes that were differentially expressed for each GO term 
y <- rbind(diffex.adult.FB %>% as.data.frame(), diffex.adult.DB %>% as.data.frame(), 
           diffex.adult %>% as.data.frame()) %>% 
  rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID, Note), by = c("gene"="Name")) %>%
    right_join(all.go.enriched %>% separate_rows(genes), by=c("SPID"="genes"))

n <- 44
z <- y %>% filter(go==all.go.enriched[n,]$go) %>% filter(group=="All Cohorts")
print(paste("GO term =", z$go %>% unique()))
genes_plots <- list()
for (i in 1:length(z$gene)) {
print(i)
a <- plotCounts(dds.adult.DESeq, gene=z$gene[i], 
             intgroup=c("population", "pCO2.parent"), returnData = TRUE) %>%rownames_to_column("sample")
b <-  ggplot(a, aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
    geom_point(position=position_jitter(w = 0.1,h = 0)) +
      geom_text() +
      theme_bw() +
      ggtitle(z$gene[i]) +
      theme(plot.title = element_text(hjust = 0.5))
genes_plots[[i]] <- b
}
genes_plots
```

# Here plot specific genes 

```{r}
# First make a dataframe with ALL data annotated 
all <- res.adult.FB %>% as.data.frame() %>% 
  rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID, Note), by = c("gene"="Name")) %>%
    left_join(all.go.enriched %>% separate_rows(genes), by=c("SPID"="genes"))

g <- c("A2AX52")
z <- all %>% filter(SPID %in% g) %>% dplyr::select(gene, SPID, Note) %>% unique()
#z <- all %>% filter(grepl("roton channel", Note)) %>% dplyr::select(gene, SPID, Note) %>% unique()
genes_plots <- list()
for (i in 1:nrow(z)) {
print(i)
a <- plotCounts(dds.adult.DESeq, gene=z$gene[i], 
             intgroup=c("population", "pCO2.parent"), returnData = TRUE) %>%rownames_to_column("sample")
b <-  ggplot(a, aes(x=population:pCO2.parent, y=count, color=population:pCO2.parent, label=sample)) +
    #geom_point(position=position_jitter(w = 0.1,h = 0)) +
  geom_boxplot() +
      geom_jitter() +
      theme_bw() +
      ggtitle(z$Note[i]) +
      theme(plot.title = element_text(hjust = 0.5))
genes_plots[[i]] <- b
}
genes_plots
ggplotly(genes_plots[[5]])
```


```{r}
#install.packages("scatterpie")
#require(scatterpie)

# Plot enriched Biological Processes 
pdf(file="../results/EnrichedBP-pCO2.pdf", height = 6.5, width = 6.5)
read.csv(file="../results/Adults.go.enriched.4paper-edited.csv", header = T)[,-1] %>% 
  mutate(go=as.factor(go)) %>% filter(GO.category=="Biological Processes") %>%
  mutate(Abundance.HighPCO2=factor(Abundance.HighPCO2, levels=c("More", "Both", "Less"))) %>%
  mutate(group=fct_relevel(as.factor(group), "All Cohorts", after = Inf)) %>% 
  filter(group!="All Cohorts") %>% 
  ggplot(aes(y=str_wrap(process.simple),x=group)) + 
  facet_wrap(~GO.category,scales="free", nrow = 2) +
  geom_point(aes(alpha=fdr,size=count))+  #size=100*perc_total  #col=Abundance.HighPCO2
  scale_x_discrete(labels=c("Dabob Bay"="Dabob\nBay", "Fidalgo Bay"="Fidalgo\nBay", "Oyster Bay"="Oyster\nBay", "All Cohorts"="All\nAdults")) +
  scale_color_manual(name="Transcript abundance\nin High pCO2",na.translate=FALSE,
                     labels=c("Greater", "Mixed", "Fewer"),values=c("royalblue3", "seagreen", "red"), guide = guide_legend(override.aes = list(size=4)))+
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.3), breaks=c(0.05, 0.25, 0.75, 1), guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("No. of genes", range = c(3,10), breaks = c(2, 5, 10), guide = guide_legend(override.aes = list(col="gray50"))) + 
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), axis.text.x = element_text(size=8),
        legend.position = "right", axis.text.y=element_text(size=8)) +
  ylab("Enriched Gene Ontology (GO) Term")
dev.off()

# Plot enriched Molecular Functions
pdf(file="../results/EnrichedMF-pCO2.pdf", height = 6.25, width = 6.9)
read.csv(file="../results/Adults.go.enriched.4paper-edited.csv", header = T)[,-1] %>% 
  mutate(go=as.factor(go)) %>% filter(GO.category=="Molecular Functions") %>%
  mutate(Abundance.HighPCO2=factor(Abundance.HighPCO2, levels=c("More", "Both", "Less"))) %>%
  mutate(group=fct_relevel(as.factor(group), "All Cohorts", after = Inf)) %>% 
  #filter(group!="All Cohorts") %>%
  ggplot(aes(y=str_wrap(process.simple),x=group))+ 
  geom_point(aes(alpha=fdr,size=count))+ #, col=Abundance.HighPCO2
 facet_wrap(~GO.category,scales="free", nrow = 2) +
    scale_x_discrete(labels=c("Dabob Bay"="Dabob\nBay", "Fidalgo Bay"="Fidalgo\nBay", "Oyster Bay"="Oyster\nBay", "All Cohorts"="All\nCohorts")) +
  scale_color_manual(name="Transcript abundance\nin High pCO2", na.translate=FALSE,
                     labels=c("Greater", "Mixed", "Fewer"),values=c("royalblue3", "seagreen", "red"), guide = guide_legend(override.aes = list(size=4)))+
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.3), guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("No. of genes", range = c(3,9), breaks = c(2, 5, 10), guide = guide_legend(override.aes = list(col="gray50"))) + 
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), axis.text.x = element_text(size=8),
        legend.position = "right", axis.text.y=element_text(size=8)) +
  ylab("Enriched Gene Ontology (GO) Term")
dev.off()


# playing around - here i plot pie charts showing no. of genes with more (blue) and less (red) transcripts in high pCO2 

read.csv(file="../results/Adults.go.enriched.4paper-edited.csv", header = T)[,-1] %>% 
  mutate(go=as.factor(go)) %>% #filter(GO.category=="Biological Processes") %>%
  mutate(Abundance.HighPCO2=factor(Abundance.HighPCO2, levels=c("More", "Both", "Less"))) %>%
  filter(group!="All Cohorts") %>% 
  pivot_longer(names_to="up.down", values_to="count.up.down", cols = c("count_up", "count_down")) %>%
  filter(Abundance.HighPCO2=="Both") %>%
  ggplot() +
  geom_bar(aes(x = "", y = count.up.down, fill = up.down, alpha=fdr, group=group),
           width = 1, stat = "identity", color = "white", position = position_fill()) +
  coord_polar("y", start = 0)+
  geom_text(aes(x = "", y = count.up.down, label = count.up.down), position = position_fill(vjust = 0.5), size=2.5) +
  scale_fill_manual(values = c("red", "royalblue3")) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.3), guide = guide_legend(override.aes = list(size=4))) +
  theme_void() +
  facet_wrap(~group+str_wrap(process.simple, width = 20), nrow = 2) +
  theme(text = element_text(size=8))
```
# Here I view all DEGs and their associated GO terms 

```{r}
rbind(diffex.adult.FB %>% as.data.frame(), diffex.adult.DB %>% as.data.frame(), 
           diffex.adult %>% as.data.frame()) %>% 
  rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID, Note, Ontology_term), by = c("gene"="Name")) %>%
    left_join(all.go.enriched %>% separate_rows(genes), by=c("SPID"="genes")) %>% View()
```

# Here I save all annotated genes that are DEGs within a cohort 

```{r}
rbind(diffex.adult.FB %>% as.data.frame() %>% mutate(population="Fidalgo Bay")) %>% 
  rownames_to_column("gene") %>% dplyr::select(gene,log2FoldChange,padj,population) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID, Note), by = c("gene"="Name")) %>% filter(!is.na(SPID)) %>%
  mutate(Note=gsub("Note=Similar to ", "", Note)) %>%
  mutate(Note=gsub("\\(.*", "", Note)) %>%
  write_csv(file = "../results/PAPER-DEGs-Fidalgo-Bay.csv")

rbind(diffex.adult.DB %>% as.data.frame() %>% mutate(population="Dabob Bay")) %>% 
  rownames_to_column("gene") %>% dplyr::select(gene,log2FoldChange,padj,population) %>% 
  left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID, Note), by = c("gene"="Name")) %>% filter(!is.na(SPID)) %>%
  mutate(Note=gsub("Note=Similar to ", "", Note)) %>%
  mutate(Note=gsub("\\(.*", "", Note)) %>%
  write_csv(file = "../results/PAPER-DEGs-Dabob-Bay.csv")

Reduce(intersect, list(rownames(diffex.adult.DB),rownames(diffex.adult.FB)))
```


# For each GO term, assess whether counts were MORE abundant or LESS abundant in each cohort! 

```{r}
EnrBP.adult.DBFB <- read_delim(file="../results/Adult_cohort_DEGs_DBFB_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Dabob Bay\n~Fidalgo Bay", GO.category="Biological Processes")

EnrMF.adult.DBFB <- read_delim(file="../results/Adult_cohort_DEGs_DBFB_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Dabob Bay\n~Fidalgo Bay", GO.category="Molecular Function")

EnrBP.adult.DBOB1 <- read_delim(file="../results/Adult_cohort_DEGs_DBOB1_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Dabob Bay\n~Oyster Bay", GO.category="Biological Processes")

EnrMF.adult.DBOB1 <- read_delim(file="../results/Adult_cohort_DEGs_DBOB1_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Dabob Bay\n~Oyster Bay", GO.category="Molecular Function")

EnrBP.adult.FBOB1 <- read_delim(file="../results/Adult_cohort_DEGs_FBOB1_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Fidalgo Bay\n~Oyster Bay", GO.category="Biological Processes")

EnrMF.adult.FBOB1 <- read_delim(file="../results/Adult_cohort_DEGs_FBOB1_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Fidalgo Bay\n~Oyster Bay", GO.category="Molecular Function")

all.go.enriched.cohorts.adults <- rbind(
  EnrBP.adult.DBFB,
  EnrMF.adult.DBFB,
  EnrBP.adult.DBOB1,
  EnrMF.adult.DBOB1,
  EnrBP.adult.FBOB1,
  EnrMF.adult.FBOB1) %>% 
  left_join(GOslim, by=c("go"="GO_ID"))

  
save(all.go.enriched.cohorts.adults, file = "../results/all.go.enriched.cohorts.adults")

# Plot enriched Biological Processes 
all.go.enriched.cohorts.adults %>% 
  mutate(go=as.factor(go)) %>% filter(GO.category=="Biological Processes") %>%
  ggplot(aes(y=str_wrap(process),x=contrast)) +  # NOTE: could instead use the next line to show GO Term ID's on the y-axis 
  geom_point(aes(alpha=fdr,size=count))+
 facet_wrap(~GO.category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.3), breaks = c(.25, .5, .75, 1), guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,12), breaks = c(2, 5, 10), guide = guide_legend(override.aes = list(col="gray50"))) + 
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), axis.text.x = element_text(size=7),
        legend.position = "right", axis.text.y=element_text(size=8.5), plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + ggtitle("Enriched function among cohorts, Adults")

# Plot enriched Molecular Functions
all.go.enriched.cohorts.adults %>% 
  mutate(go=as.factor(go)) %>% filter(GO.category=="Molecular Function") %>%
  ggplot(aes(y=str_wrap(process),x=contrast))+  # NOTE: could instead use the next line to show GO Term ID's on the y-axis 
  geom_point(aes(alpha=fdr,size=count)) +
 facet_wrap(~GO.category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.25), breaks = c(.25, .5, .75, 1), guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,12), breaks = c(2, 5, 10), guide = guide_legend(override.aes = list(col="gray50"))) + 
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), axis.text.x = element_text(size=7),
        legend.position = "right", axis.text.y=element_text(size=8.5), plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + ggtitle("Enriched function among cohorts, Adults")
```


# PCA with all genes using princomp and showing broken stick 

```{r}

# PCA using prcomp etc., but NOTE that I pull the covariance matrix from the SNPRelate PCA command  
pca.adult.princomp.expr <- prcomp(cov(assay(vsd.adult)), scale=F) #scale=F for variance-covariance matrix
pca.eigenval(pca.adult.princomp.expr) #The Proporation of Variance = %variance 
pc.percent.adult <- pca.eigenval(pca.adult.princomp.expr)[2,1:4]*100
screeplot(pca.adult.princomp.expr, bstick=TRUE) #looks like PC 1-5 are significant 

tab.adult.expr <- data.frame(sample.id = colnames(assay(vsd.adult)),
    EV1.all = pca.adult.princomp.expr$x[,1],    # the first eigenvector
    EV2.all = pca.adult.princomp.expr$x[,2],    # the second eigenvector
    stringsAsFactors = FALSE)

shapiro.test(pca.adult.princomp.expr$x) #multivariate normality met 

tab.adult.expr.annot <- left_join(tab.adult.expr, sample.info.adult, by=c("sample.id"="sample")) %>% droplevels()
group.adult.expr <- as.factor(tab.adult.expr.annot$pop_code)

ggplot(tab.adult.expr.annot, aes(x=EV1.all, y=EV2.all)) + geom_point(aes(shape=population, col=pCO2.parent), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Adult Expression, all genes, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.adult[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.adult[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(values=c("#4575b4","#d73027"), name="Adult pCO2 Exposure") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=pCO2.parent), size=0.3)

ggplot(tab.adult.expr.annot, aes(x=EV1.all, y=EV2.all)) + geom_point(aes(col=population, shape=pCO2.parent), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Adult Expression, all genes, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.adult[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.adult[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(values=c("#4daf4a","#377eb8","#984ea3"), name="Cohort") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=population, linetype=pCO2.parent), size=0.5)

save(tab.adult.expr.annot, file="../results/tab.adult.expr.annot")

# GGSCATTER with ellipses and stars 
ggscatter(tab.adult.expr.annot, x="EV1.all", y="EV2.all",
          shape="population", color="pCO2.parent", size=3.5, alpha=0.85, 
           palette = c("#4575b4","#d73027"),
   ellipse = TRUE, star.plot = TRUE) + 
  theme_minimal() + ggtitle("Adult Expression, all genes, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.adult[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.adult[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  guides(colour = guide_legend(override.aes = list(size=3.5))) 
```


# PCA with gene set that are DEGs among any pCO2 comparison (all pops, FB, DB, and OB1)

```{r}
pca.adult.princomp.expr.co2 <-
  princomp(cov(assay(vsd.adult[unique(c(rownames(diffex.adult),rownames(diffex.adult.FB),rownames(diffex.adult.DB),rownames(diffex.adult.OB1))),])))
pca.eigenval(pca.adult.princomp.expr.co2) #The Proporation of Variance = %variance 
pc.percent.adult.pco2 <- pca.eigenval(pca.adult.princomp.expr.co2)[2,1:4]*100
screeplot(pca.adult.princomp.expr.co2, bstick=TRUE) #looks like PC 1 & 2

tab.adult.expr.co2 <- data.frame(sample.id = colnames(assay(vsd.adult)),
    EV1.pco2 = pca.adult.princomp.expr.co2$scores[,1],    # the first eigenvector
    EV2.pco2 = pca.adult.princomp.expr.co2$scores[,2],    # the second eigenvector
    stringsAsFactors = FALSE)

tab.adult.expr.pco2.annot <- left_join(tab.adult.expr.co2, sample.info.adult, by=c("sample.id"="sample")) %>% droplevels()
group.adult.pco2.expr <- as.factor(tab.adult.expr.pco2.annot$pop_code)

ggplot(tab.adult.expr.pco2.annot, aes(x=EV1.pco2, y=EV2.pco2)) + geom_point(aes(shape=population, col=pCO2.parent), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Adult Expression PC1xPC2, DEGs in response to pCO2") + 
  xlab(paste("PC1 (", round(pc.percent.adult.pco2[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.adult.pco2[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(values=c("#4575b4","#d73027"), name="Adult pCO2 Exposure") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=pCO2.parent), size=0.3)

save(tab.adult.expr.pco2.annot, file="../results/tab.adult.expr.pco2.annot")
```

# PCA with gene set that are DEGs among any pairwise population comparison 

```{r}
pca.adult.princomp.expr.pop <-
  princomp(cov(assay(vsd.adult.ambient[degs.adult.pops.amb,])))
pca.eigenval(pca.adult.princomp.expr.pop) #The Proporation of Variance = %variance 
pc.percent.adult.pop <- pca.eigenval(pca.adult.princomp.expr.pop)[2,1:4]*100
screeplot(pca.adult.princomp.expr.pop, bstick=TRUE) #looks like PC 1, 2 are sign.

tab.adult.expr.pop <- data.frame(sample.id = colnames(assay(vsd.adult.ambient)),
    EV1.coh = pca.adult.princomp.expr.pop$scores[,1],    # the first eigenvector
    EV2.coh = pca.adult.princomp.expr.pop$scores[,2],    # the second eigenvector
    stringsAsFactors = FALSE)

tab.adult.expr.pop.annot <- left_join(tab.adult.expr.pop, sample.info.adult, by=c("sample.id"="sample")) %>% droplevels()
group.adult.pop.expr <- as.factor(tab.adult.expr.pop.annot$pop_code)

ggplot(tab.adult.expr.pop.annot, aes(x=EV1.coh, y=EV2.coh)) + geom_point(aes(shape=pCO2.parent, col=population), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Adult Expression PC1xPC2, DEGs among cohorts") + 
  xlab(paste("PC1 (", round(pc.percent.adult.pop[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.adult.pop[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(values=c("#4daf4a","#377eb8","#984ea3"), name="Cohort") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=population), size=0.3)
save(tab.adult.expr.pop.annot, file="../results/tab.adult.expr.pop.annot")
```
Heatmap of genes involved in oxidation reduction processes 

```{r}
annotation_colors = list(
  pCO2.parent = c(Ambient="royalblue3", High="red"),
  population=c(`Dabob Bay`="#4daf4a", `Fidalgo Bay`="#377eb8", `Oyster Bay C1`="#984ea3"))

# all <- rbind(res.adult.FB %>% as.data.frame()) %>% 
#   rownames_to_column("gene") %>% dplyr::select(gene, padj) %>% 
#   left_join(Olurida_gene_uniprot %>% dplyr::select(Name, SPID, Note), by = c("gene"="Name")) %>%
#     left_join(all.go.enriched %>% separate_rows(genes), by=c("SPID"="genes"))

# heatmap of genes involved in detoxification, oxidation reduction stuff 
oxid.adult.counts <- subset(counts(dds.adult.DESeq), rownames(dds.adult.DESeq) %in% 
                              c(unique((y %>% filter(grepl("oxid|sulfo", process)))$gene), 
                                                       (y %>% filter(grepl("crystallin", Note)))$gene))

oxid.adult.counts <- oxid.adult.counts[,(key.filtered %>% filter(stage=="adult") %>% arrange(population, pCO2.parent))$sample]
dds.adult.df.oxid <- dds.adult.df[match(colnames(oxid.adult.counts), rownames(dds.adult.df)),]
all(colnames(oxid.adult.counts) == rownames(dds.adult.df.oxid)) #double check that samples are still in same order 

pheatmap(log2(oxid.adult.counts+1), cluster_rows=T, cluster_cols = F,
         scale="row", show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, 
         main = "Adult DEGs involved in detoxification, redox", 
         annotation_col=(dds.adult.df %>% rename("population"="Cohort"))[1:2], 
         annotation_colors = list(pCO2.parent = c(Ambient="royalblue3", High="red"),
  Cohort=c(`Dabob Bay`="#4daf4a", `Fidalgo Bay`="#377eb8", `Oyster Bay C1`="#984ea3")), fontsize = 8.5)

y %>% filter(gene %in% rownames(oxid.adult.counts)) %>% distinct(SPID, group, .keep_all = TRUE)
y %>% filter(gene %in% rownames(oxid.adult.counts)) %>% distinct(gene, .keep_all = TRUE) %>% View()

# Save table of detox/redox genes FOR PAPER 
y %>% filter(gene %in% rownames(oxid.adult.counts)) %>% distinct(SPID, .keep_all = TRUE) %>% select(gene, padj, SPID, Note, group) %>%
  write_csv(file="../results/PAPER-detox-redox-DEGs.csv", col_names = TRUE)


# heatmap of genes involved in lipid transport
lipid.adult.counts <- subset(counts(dds.adult.DESeq), rownames(dds.adult.DESeq) %in% 
                                 c(unique(y %>% filter(grepl("Abc|ABC|phospholipase|von Willebrand", Note)))$gene,
                                unique(y %>% filter(grepl("phospholipid translocation|lipid transport|fatty|lipid", process)))$gene))
# %>% filter(!grepl("Vitellogenin|Cytosolic phospholipase A2|HELZ2: Helicase with zinc"))

lipid.adult.counts <- lipid.adult.counts[,(key.filtered %>% filter(stage=="adult") %>% arrange(population, pCO2.parent))$sample]
dds.adult.df.lipid <- dds.adult.df[match(colnames(lipid.adult.counts), rownames(dds.adult.df)),]
all(colnames(lipid.adult.counts) == rownames(dds.adult.df.lipid)) #double check that samples are still in same order 

pheatmap(log2(lipid.adult.counts[!rownames(lipid.adult.counts) %in% c("OLUR_00007563", "OLUR_00001118", "OLUR_00013228"),]+1), 
         cluster_rows=T, cluster_cols = F,
         scale="row", show_rownames=T, show_colnames = FALSE, na.rm=TRUE, 
         main = "Adult DEGs involved in lipid metabolism and transport", 
         annotation_col=dds.adult.df[1:2], annotation_colors = annotation_colors, fontsize = 8.5)


# boxplots of stress-related genes 

g <- (all %>% filter(grepl("heat shock|Heat shock|Universal stress|Stress response|Stress-activated|Stress-induced", Note)))$gene
z <- all %>% filter(gene %in% g) %>% dplyr::select(gene, SPID, Note) %>% unique()
z <- all %>% filter(gene %in% rownames(lipid.adult.counts)) %>% dplyr::select(gene, SPID, Note) %>% unique()
genes_plots <- list()
for (i in 1:nrow(z)) {
print(i)
a <- plotCounts(dds.adult.DESeq, gene=z$gene[i], 
             intgroup=c("population", "pCO2.parent"), returnData = TRUE) %>% rownames_to_column("sample")
b <-  ggplot(a, aes(x=population:pCO2.parent, y=count, color=population:pCO2.parent, label=sample)) +
    #geom_point(position=position_jitter(w = 0.1,h = 0)) +
  geom_boxplot() +
      geom_jitter() +
      theme_bw() +
      ggtitle(z$Note[i]) +
      theme(plot.title = element_text(hjust = 0.5))
genes_plots[[i]] <- b
}

genes_plots
#ggplotly(genes_plots[[1]])


# heatmap of genes involved in channels
channel.adult.counts <- subset(counts(dds.adult.DESeq), rownames(dds.adult.DESeq) %in% 
                                 c(unique(all %>% filter(grepl("hydrogen channel|proton", Note)))$gene))

channel.adult.counts <- channel.adult.counts[,(key.filtered %>% filter(stage=="adult") %>% arrange(population, pCO2.parent))$sample]
dds.adult.df.channel <- dds.adult.df[match(colnames(channel.adult.counts), rownames(dds.adult.df)),]
all(colnames(channel.adult.counts) == rownames(dds.adult.df.channel)) #double check that samples are still in same order 


pheatmap(log2(channel.adult.counts+1), 
         cluster_rows=T, cluster_cols = F,
         scale="row", show_rownames=T, show_colnames = FALSE, na.rm=TRUE, 
         main = "Adult genes (not DEGs) involved in proton channels", 
         annotation_col=dds.adult.df[1:2], annotation_colors = annotation_colors, fontsize = 8.5)

dds.adult.df.oxid %>% group_by(population,pCO2.parent) %>% tally()
```

# Plot stress-response genes 

```{r}
g <- (y %>% filter(grepl("Q99933|O43301|P74897|Q9D4G2", SPID)))$gene
z <- all %>% filter(gene %in% g) %>% dplyr::select(gene, SPID, Note) %>% unique() %>%
  mutate(protein=c("BAG1 (molecular chaperone)", "HSPA12A", "Universal stress protein in\nQAH/OAS sulfhydrylase 3'region", "Hsf2bp"))

count_plots.adult = list()
for (i in 1:nrow(z)) {
a <-  plotCounts(dds.adult.DESeq, gene=z$gene[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
b <- ggplot(a %>% rownames_to_column("sample"),
       aes(x=population:pCO2.parent, y=count, color=population, shape=pCO2.parent, label=sample)) +
  geom_point(size=3, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text(size=3) +
    theme_bw() +
    ggtitle(z$protein[i]) +
    theme(plot.title = element_text(hjust = 0.5, size = 10), axis.text.x = element_blank(),
          axis.text.y = element_text(size=6),
          axis.title = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
  scale_color_manual(values = c("#4daf4a","#377eb8","#984ea3"), name="Cohort", guide = guide_legend(override.aes = list(size=4))) +
  scale_shape(name="pCO2 exposure", guide = guide_legend(override.aes = list(size=4))) 
count_plots.adult[[i]] <- b
}
plot_grid(plotlist=count_plots.adult) #plot them in a grid 
```

# Plot glutathione genes 

```{r}
g <- (all %>% filter(grepl("lutathione", Note)))$gene
z <- all %>% filter(gene %in% g) %>% dplyr::select(gene, SPID, Note) %>% unique() 

count_plots.adult = list()
for (i in 1:nrow(z)) {
a <-  plotCounts(dds.adult.DESeq, gene=z$gene[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
b <- ggplot(a %>% rownames_to_column("sample"),
       aes(x=population:pCO2.parent, y=count, color=population, shape=pCO2.parent, label=sample)) +
  geom_point(size=3, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text(size=3) +
    theme_bw() +
    ggtitle(z$SPID[i]) +
    theme(plot.title = element_text(hjust = 0.5, size = 10), axis.text.x = element_blank(),
          axis.text.y = element_text(size=6),
          axis.title = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
  scale_color_manual(values = c("#4daf4a","#377eb8","#984ea3"), name="Cohort", guide = guide_legend(override.aes = list(size=4))) +
  scale_shape(name="pCO2 exposure", guide = guide_legend(override.aes = list(size=4))) 
count_plots.adult[[i]] <- b
}
plot_grid(plotlist=count_plots.adult) #plot them in a grid 
```

# Boxplot of oxidation reduction genes for PAPER 

```{r}
left_join(t(oxid.adult.counts) %>% as.data.frame() %>% rownames_to_column(var="sample"),
          dds.adult.df.oxid %>% rownames_to_column(var = "sample"),
          by = "sample") %>%
  group_by(population, pCO2.parent) %>%
  summarise_if(is.numeric, mean) %>%
  pivot_longer(cols = c(-population, -pCO2.parent)) %>%
  mutate(pCO2.parent=gsub("Ambient", "Control", pCO2.parent)) %>%
  mutate(pCO2.parent=gsub("High", "Acidification", pCO2.parent)) %>%
  mutate(population=gsub("Oyster Bay C1", "Oyster Bay", population)) %>%
  
  ggplot(aes(x=fct_rev(pCO2.parent), y=value, fill=fct_rev(pCO2.parent))) + 
  geom_boxplot(alpha=0.55) +
  #geom_violin() +
  geom_dotplot(binaxis='y', stackdir='center', alpha=0.3) + 
  #geom_line(aes(group=name)) +
  facet_wrap(~population) +
  scale_fill_manual(values=c("blue","red"), name="Treatment") +
  # geom_point(data=b, aes(x=fct_rev(time), y=mean, fill=pH, group=pH), color="black", size=4.5, pch=21) + 
  # geom_line(data=b, aes(x=fct_rev(time), y=mean, color=pH, group=pH)) + 
  scale_color_manual(values=c("#67a9cf","#ef8a62"), guide="none") +
  theme_bw(base_size = 12) + 
  theme(plot.title = element_text(size = 13, hjust = 0, colour = "gray30"), panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), panel.background = element_blank(), 
        axis.title.x = element_blank(), legend.position="right", axis.text.x = element_text(size=8)) +  
          labs(title=(expression(paste("Normalized counts for genes involved in detoxification, redox"))),  # by ", pCO[2], " exposure & cohort
               y=expression("Mean transcript abundance")) +
      stat_summary(fun.y=mean, geom="point", shape=22, size=3, color="black", fill="black")

#show_col(c('#d1e5f0','#67a9cf','#fddbc7','#ef8a62'))
  
```
# bubble plot for FINAL EXAM not including ALL populations 
```{r}
# copy 650 x 650
read.csv(file="results/Adults.go.enriched.4paper-edited.csv", header = T)[,-1] %>% 
  mutate(go=as.factor(go)) %>% filter(GO.category=="Biological Processes" & group!="All Cohorts") %>%
  mutate(Abundance.HighPCO2=factor(Abundance.HighPCO2, levels=c("More", "Both", "Less"))) %>%
  #mutate(group=fct_relevel(as.factor(group), "All Cohorts", after = Inf)) %>% 
  #filter(group!="All Cohorts") %>% 
  ggplot(aes(y=str_wrap(process.simple),x=group)) + 
  facet_wrap(~GO.category,scales="free", nrow = 2) +
  geom_point(aes(size=count), alpha=0.75, color="gray50")+  #alpha=fdr, col=Abundance.HighPCO2
  scale_x_discrete(labels=c("Dabob Bay"="Dabob\nBay", "Fidalgo Bay"="Fidalgo\nBay", "Oyster Bay"="Oyster\nBay", "All Cohorts"="All\nCohorts")) +
  #scale_color_manual(name="Transcript abundance\nin High pCO2",na.translate=FALSE,
  #                   labels=c("Greater", "Mixed", "Fewer"),values=c("royalblue3", "seagreen", "red"), guide = guide_legend(override.aes = list(size=4)))+
  #scale_alpha("FDR-adjusted\nP-value", range = c(1,0.3), breaks=c(0.05, 0.25, 0.75, 1), guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("No. of genes", range = c(3,10), breaks = c(2, 5, 10), guide = guide_legend(override.aes = list(col="gray50"))) + 
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), axis.text.x = element_text(size=8),
        legend.position = "right", axis.text.y=element_text(size=8)) +
  ylab("Enriched Gene Ontology (GO) Term")
```

# WHAT'S DIFFERENT ABOUT DABOB BAY? 

```{r}
# generate heatmap of genes diff. expressed among Dabob Bay & other pops   
degs.adult.dabob.amb <- c(rownames(diffex.adult.DBFB.amb), 
  rownames(diffex.adult.DBOB1.amb)) %>% unique()

rownames(diffex.adult.DBFB.amb) %>% length()
rownames(diffex.adult.DBOB1.amb) %>% length()

# This is how many genes were DEGs among Dabob Bay and both of the other populations 
Reduce(intersect, list(rownames(diffex.adult.DBFB.amb),rownames(diffex.adult.DBOB1.amb))) %>% length()

# What do those 89 genes do? 33 are annotated
filter(Olurida_gene_uniprot, 
       Name %in% Reduce(intersect, list(rownames(diffex.adult.DBFB.amb),rownames(diffex.adult.DBOB1.amb)))) %>%
  filter(!is.na(SPID))

print(out.adult.dabob <- pheatmap(assay(vsd.adult.ambient[degs.adult.dabob.amb,]), cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[3], scale="row", main = "DEGs among Dabob Bay & other pops\nAdults (ambient pCO2 only)",
         annotation_colors=list(Cohort=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")), 
         cutree_cols = 3))
```

```{r}
# extract gene clusters 
hc.adult.dabob <-out.adult.dabob$tree_row

# identify which "height" to cut tree for 2 clusters 
plot(hc.adult.dabob)
abline(h=9.5, col="red", lty=2, lwd=2)  

lbl.adult.dabob <- cutree(hc.adult.dabob, h=9.5) # split gene dendrogram into 2 clusters, alternatively could use "k=2" 

# Extract gene names for the  gene clusters and plot
higher.cluster.adult.dabob <- assay(vsd.adult.ambient)[which(lbl.adult.dabob==1) %>% names(),]
lower.cluster.adult.dabob <- assay(vsd.adult.ambient)[which(lbl.adult.dabob==2) %>% names(),]

# Plot each cluster separately to inspect pattern 
pheatmap(higher.cluster.adult.dabob, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row", main = "Genes more abundant in Dabob Bay",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
pheatmap(lower.cluster.adult.dabob, cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row",  main = "Genes less abundant in Dabob Bay",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))

rownames(higher.cluster.adult.dabob) %>% length() #how many are higher 
rownames(lower.cluster.adult.dabob) %>% length() #how many are lower 

318+252

# Heat map of genes higher AND lower in Dabob Bay than other populations 
pheatmap(assay(vsd.adult.ambient)[which(lbl.adult.dabob==1|2) %>% names(),], 
         cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE, cutree_rows = 2,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row", main = "Genes uniquely abundant in Dabob Bay",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))

```

```{r}
# Are any of the Dabob Bay DEGs also DEGs in response to acidification? 
DEGs.DB.adult.pHpop <- Reduce(intersect, list(c(rownames(diffex.adult.DBFB.amb),rownames(diffex.adult.DBOB1.amb)), rownames(diffex.adult.DB))) # yes! 
DEGs.DB.adult.pHpop %>% length()

# What do the genes do that were DEGs in DB in response to OA and DEGs in DB vs the other pops? 
filter(Olurida_gene_uniprot, 
       Name %in% DEGs.DB.adult.pHpop) %>% 
  filter(!is.na(SPID))

# Are the pH/pop DEGS in Dabob Bay higher or lower consitutively than the other pops? 
print(out.adult.dabob.pHpop <- pheatmap(assay(vsd.adult.ambient[DEGs.DB.adult.pHpop,]), cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[3], scale="row", main = "DEGs among Dabob Bay & other pops\nAdults (ambient pCO2 only)",
         annotation_colors=list(Cohort=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")), 
         cutree_cols = 3))

# Some are lower, some are higher. Whiech ones are which? use clustering 
# extract gene clusters 

out.adult.dabob.pHpop.tree <- cutree(out.adult.dabob.pHpop$tree_row, k=2) # split gene dendrogram into 2 clusters

# Extract gene names for the  gene clusters and plot
pheatmap(assay(vsd.adult.ambient)[which(out.adult.dabob.pHpop.tree==1) %>% names(),], 
         cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row", main = "Genes less abundant in Dabob Bay constitutively, also DEGs in OA",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
pheatmap(assay(vsd.adult.ambient)[which(out.adult.dabob.pHpop.tree==2) %>% names(),], 
         cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=F, annotation_col=vsd.adult.ambient.df[1], scale="row",  main = "Genes more abundant in Dabob Bay, also DEGs in OA",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))
```


# Enrichment analysis on each gene cluster 

```{r}
# Background list 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(assay(vsd.adult.ambient))) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# All genes uniquely abundant in Dabob Bay than other populations 
filter(Olurida_gene_uniprot, 
       Name %in% c(rownames(higher.cluster.adult.dabob),rownames(lower.cluster.adult.dabob))) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% unique() %>% length() %>% write_clip()

# the 89 genes that were DEGs among DB and both of the other populations 
filter(Olurida_gene_uniprot, 
       Name %in% Reduce(intersect, list(rownames(diffex.adult.DBFB.amb),rownames(diffex.adult.DBOB1.amb)))) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Genes HIGHER in Dabob Bay than other populations 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(higher.cluster.adult.dabob)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Genes LOWER in DB 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(lower.cluster.adult.dabob)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# Genes HIGHER in DB constitutively, and also DEGs in response to OA 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(assay(vsd.adult.ambient)[which(out.adult.dabob.pHpop.tree==2) %>% names(),])) %>% 
  filter(!is.na(SPID)) %>% mutate(Note=gsub("Note=Similar to ", "", Note)) %>% select(Note)

# Genes LOWER in DB constitutively, and also DEGs in response to OA 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(assay(vsd.adult.ambient)[which(out.adult.dabob.pHpop.tree==1) %>% names(),])) %>% 
  filter(!is.na(SPID)) %>% mutate(Note=gsub("Note=Similar to ", "", Note)) %>% select(Name)


# Saved DAVID enrichment results into the results/Dabob-Bay-Analysis/ folder 

#Save list of annotated genes to file, too 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(higher.cluster.adult.dabob)) %>% 
  filter(!is.na(SPID)) %>% dplyr::select(contig, start, end, Name, Note, Ontology_term, SPID) %>%
  write.csv(file="../results/Dabob-Bay-Analysis/higher-DB-annotated-genes.csv")

filter(Olurida_gene_uniprot, 
       Name %in% rownames(lower.cluster.adult.dabob)) %>% 
  filter(!is.na(SPID)) %>% dplyr::select(contig, start, end, Name, Note, Ontology_term, SPID) %>%
  write.csv(file="../results/Dabob-Bay-Analysis/lower-DB-annotated-genes.csv") 

# Copy all GO terms (both higher and lower activity/abundance in Dabob Bay) to paste into Revigo 
rbind(read_delim(file="../results/Dabob-Bay-Analysis/higher-DB-enriched-BP.txt", delim = "\t") %>% mutate(activity="Higher"),
read_delim(file="../results/Dabob-Bay-Analysis/lower-DB-enriched-BP.txt", delim = "\t") %>% mutate(activity="Lower")) %>%
  mutate(GO = str_extract(Term, "GO(.*?)~")) %>% 
  mutate(GO = gsub("~", "", GO)) %>% 
  dplyr::select(GO, PValue) %>% na.omit() %>% write_clip()
```

PLOT REVIGO 

```{r}
# A plotting R script produced by the Revigo server at http://revigo.irb.hr/
# If you found Revigo useful in your work, please cite the following reference:
# Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800

# --------------------------------------------------------------------------
# If you don't have the scales package installed, uncomment the following line:
# install.packages( "scales" );
library( scales );

# --------------------------------------------------------------------------
# Here is your data from Revigo. Scroll down for plot configuration options.

revigo.names <- c("term_ID","description","frequency","plot_X","plot_Y","log_size","value","uniqueness","dispensability");
revigo.data <- rbind(c("GO:0008152","metabolic process",62.6341027647714,-2.47201248913027,-7.13777004368689,7.15590104223087,-1.36095291023162,1,0),
c("GO:0010638","positive regulation of organelle organization",0.218230097871315,-5.75758760519468,-0.344265128323489,4.69801350393918,-1.71699186298277,0.83109756360054,0),
c("GO:0060716","labyrinthine layer blood vessel development",0.00311888512061271,2.77387713360372,-4.89105777132874,2.85369821177617,-1.12730080100845,0.845759959437991,0),
c("GO:0071346","cellular response to interferon-gamma",0.0126680102514648,2.3067190534451,5.34229876901363,3.46194849520376,-2.24157517007866,0.847442430294088,0),
c("GO:0098656","anion transmembrane transport",1.03128364256737,-3.51775831728212,6.50230509408839,5.37247012284187,-1.91766948999044,0.994673041427961,0.00722663),
c("GO:0000209","protein polyubiquitination",0.0898177674355411,7.00265964162933,0.357535944819143,4.31247355768606,-1.23956127132217,0.995317299879444,0.00844798),
c("GO:1901214","regulation of neuron death",0.0426932942176439,-5.48022829577577,-3.33590497031215,3.98949431277271,-1.08783603687604,0.925116004589703,0.18737541),
c("GO:0035094","response to nicotine",0.00456240838821747,3.37664822630801,4.49727599472945,3.01870049866624,-1.10333988931411,0.886078481132239,0.34641366),
c("GO:0060355","positive regulation of cell adhesion molecule production",0.000651772626888211,-6.100942800453,1.04085399254602,2.17609125905568,-1.18323914039511,0.85693202068914,0.42496586),
c("GO:0043568","positive regulation of insulin-like growth factor receptor signaling pathway",0.001644741662483,-5.10779941417564,0.529722145490226,2.57634135020579,-1.32525050116676,0.852393046308781,0.44791243),
c("GO:0048812","neuron projection morphogenesis",0.136911620462913,3.50348590273538,-4.86929606791492,4.49554433754645,-1.12437298811424,0.829777573740505,0.51273262),
c("GO:0008407","chaeta morphogenesis",0.00129042231498001,3.58738282371408,-4.18198609626961,2.47129171105894,-1.03448853886941,0.843510408357837,0.52741893),
c("GO:0071222","cellular response to lipopolysaccharide",0.0231401154109975,2.24506037083289,4.75557214016762,3.72353776153206,-1.12063483685364,0.845572722878007,0.5871506));

one.data <- data.frame(revigo.data);
names(one.data) <- revigo.names;
one.data <- one.data [(one.data$plot_X != "null" & one.data$plot_Y != "null"), ];
one.data$plot_X <- as.numeric( as.character(one.data$plot_X) );
one.data$plot_Y <- as.numeric( as.character(one.data$plot_Y) );
one.data$log_size <- as.numeric( as.character(one.data$log_size) );
one.data$value <- as.numeric( as.character(one.data$value) );
one.data$frequency <- as.numeric( as.character(one.data$frequency) );
one.data$uniqueness <- as.numeric( as.character(one.data$uniqueness) );
one.data$dispensability <- as.numeric( as.character(one.data$dispensability) );
#head(one.data);


# --------------------------------------------------------------------------
# Names of the axes, sizes of the numbers and letters, names of the columns,
# etc. can be changed below

p1 <- ggplot( data = one.data );
p1 <- p1 + geom_point( aes( plot_X, plot_Y, colour = value, size = log_size), alpha = I(0.6) ) + scale_size_area();
p1 <- p1 + scale_colour_gradientn( colours = c("blue", "green", "yellow", "red"), limits = c( min(one.data$value), 0) );
p1 <- p1 + geom_point( aes(plot_X, plot_Y, size = log_size), shape = 21, fill = "transparent", colour = I (alpha ("black", 0.6) )) + scale_size_area();
p1 <- p1 + scale_size( range=c(5, 30)) + theme_bw(); # + scale_fill_gradientn(colours = heat_hcl(7), limits = c(-300, 0) );
ex <- one.data [ one.data$dispensability < 0.15, ];
p1 <- p1 + geom_text( data = ex, aes(plot_X, plot_Y, label = description), colour = I(alpha("black", 0.85)), size = 3 );
p1 <- p1 + labs (y = "semantic space x", x = "semantic space y");
p1 <- p1 + theme(legend.key = element_blank()) ;
one.x_range = max(one.data$plot_X) - min(one.data$plot_X);
one.y_range = max(one.data$plot_Y) - min(one.data$plot_Y);
p1 <- p1 + xlim(min(one.data$plot_X)-one.x_range/10,max(one.data$plot_X)+one.x_range/10);
p1 <- p1 + ylim(min(one.data$plot_Y)-one.y_range/10,max(one.data$plot_Y)+one.y_range/10);


# --------------------------------------------------------------------------
# Output the plot to screen

p1;

# Uncomment the line below to also save the plot to a file.
# The file type depends on the extension (default=pdf).

# ggsave("/path_to_your_file/revigo-plot.pdf");
```

# How do genes that are constitutively expressed differently in Dabob AND are OA-DEGs respond to OA? 

```{r}
# Genes HIGHER in DB constitutively, how do they respond to OA? 
DB.higherpop.pH <- rownames(assay(vsd.adult.ambient)[which(out.adult.dabob.pHpop.tree==2) %>% names(),])

# #plot DB DEGs to find any issues 
count_plots <- list()
for (i in 1:length(DB.higherpop.pH)) {
a <-  plotCounts(dds.adult.DESeq, gene=DB.higherpop.pH[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
b <- ggplot(a %>% rownames_to_column("sample"),
       aes(x=population:pCO2.parent, y=count, color=population:pCO2.parent, label=sample)) +
  geom_point(position=position_jitter(w = 0.1,h = 0)) +
    geom_text() +
    theme_bw() +
    ggtitle(DB.higherpop.pH[i]) +
    theme(plot.title = element_text(hjust = 0.5))
count_plots[[i]] <- b
}
count_plots
                                  

# Genes LOWER in DB constitutively, how do they respond to OA? 
DB.lowerpop.pH <- rownames(assay(vsd.adult.ambient)[which(out.adult.dabob.pHpop.tree==1) %>% names(),])

# #plot DB DEGs to find any issues 
count_plots <- list()
for (i in 1:length(DB.lowerpop.pH)) {
a <-  plotCounts(dds.adult.DESeq, gene=DB.lowerpop.pH[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
b <- ggplot(a %>% rownames_to_column("sample"),
       aes(x=population:pCO2.parent, y=count, color=population:pCO2.parent, label=sample)) +
  geom_point(position=position_jitter(w = 0.1,h = 0)) +
    geom_text() +
    theme_bw() +
    ggtitle(DB.lowerpop.pH[i]) +
    theme(plot.title = element_text(hjust = 0.5))
count_plots[[i]] <- b
}
count_plots

```

# New Analysis - September 2021 

Determine relative/rank abundances of genes involved in specific processes (e.g. stress-response, lipid metabolism), 0

```{r}
# Figure out which gene has highest abundance in each sample - 
test <- counts.tsk.adult[1:6,grepl("OLUR", colnames(counts.tsk.adult))] %>% t() %>% as.data.frame()

test2 <- counts.tsk.adult[1:6,grepl("OLUR", colnames(counts.tsk.adult))] %>% t() %>% as.data.frame() %>% mutate_each(funs(percent_rank(-.)))

# Check to see if the most abundant gene in the first 6 samples is also ranked the highest
all(
    c(test[which.max(test$`301`),] %>% rownames(),
test[which.max(test$`302`),] %>% rownames(),
test[which.max(test$`303`),] %>% rownames(),
test[which.max(test$`304`),] %>% rownames(),
test[which.max(test$`305`),] %>% rownames(),
test[which.max(test$`306`),] %>% rownames()) == 

  c(test2[which.min(test2$`301`),] %>% rownames(),
test2[which.min(test2$`302`),] %>% rownames(),
test2[which.min(test2$`303`),] %>% rownames(),
test2[which.min(test2$`304`),] %>% rownames(),
test2[which.min(test2$`305`),] %>% rownames(),
test2[which.min(test2$`306`),] %>% rownames()))
# Yep - code for "test2" is good. 

# Determine rank of each gene within each sample, rescaled to 0-1 
ranks.adult <- counts.tsk.adult[grepl("OLUR", colnames(counts.tsk.adult))] %>% t() %>% as.data.frame() %>% 
  mutate_each(funs(percent_rank(-.))) %>% t() %>% as.data.frame() %>% rownames_to_column("sample") %>%
  left_join(counts.tsk.adult[!grepl("OLUR", colnames(counts.tsk.adult))] %>% rownames_to_column("sample"), .) %>%
   pivot_longer(cols=starts_with("OLUR"), names_to="gene", values_to="rank")

# Annotate genes with GO terms and GO Slim 
ranks.adult.annot <- left_join(ranks.adult[,c("sample","population","pCO2.parent","gene","rank")], 
                               Olurida_gene_uniprot[,c("Name", "Ontology_term", "SPID", "Note")], by=c("gene"="Name")) %>%
  mutate(Ontology_term=gsub("Ontology_term=", "", Ontology_term)) %>% mutate(Ontology_term=gsub(";", "", Ontology_term)) %>%
  separate_rows(Ontology_term, sep=",") %>% left_join(., GOslim, by=c("Ontology_term"="GO_ID")) %>%
  mutate_each_(funs(factor(.)),c("sample", "population", "pCO2.parent", "gene", "term", "GOSlim_bin"))

# Plot relative rank of genes related to stress response by population & pCO2 (each sample has its own line)
#ggplotly(
ranks.adult.annot %>% 
  filter(GOSlim_bin=="stress response") %>% 
  #filter(term=="response to heat") %>%
  ggplot(aes(x=pCO2.parent, y=rank, group=gene, text=Note)) + geom_point(size=0.1) + 
  stat_summary(aes(group = gene), geom = "line", fun.y = mean) +
  theme_minimal() + theme(legend.position="none") + 
    facet_wrap(~population) #) 

# Make new dataframe with average rank per gene for each population + pCO2 treatment 
ranks.adult.annot.ave <- ranks.adult.annot %>% 
  group_by(population, pCO2.parent, gene, Ontology_term, SPID, Note, term, GOSlim_bin) %>% 
  summarize(mean_rank = mean(rank))

# Plot distribution of mean relative rank 
ranks.adult.annot.ave %>% ggplot(aes(mean_rank)) + geom_histogram() + facet_wrap(~pCO2.parent+population)

# Plot mean relative rank of genes related to stress response by population & pCO2 
ggplotly(
ranks.adult.annot.ave %>% 
  filter(GOSlim_bin=="stress response") %>% 
  #filter(term=="oxidoreductase activity") %>%
  ggplot(aes(x=pCO2.parent, y=mean_rank, group=gene, text=Note, col=gene)) + geom_line() + 
  theme_minimal() + theme(legend.position="none") + 
    facet_wrap(~population)) #  

# Calculate mean slope 
# Calculate mean height beginnin and end 
# Which genes are already higher in Dabob Bay and don't move or go up a little -what do they do?  
# Calculate mean slope for all genes, then identify those with negative slope and those with positive slope 
# Keep concept of "plasticity" in mind ... 

# Plot mean relative rank of genes related to mitochondrion by population & pCO2 
ggplotly(ranks.adult.annot.ave %>% filter(GOSlim_bin=="mitochondrion") %>% 
  ggplot() + geom_line(aes(x=pCO2.parent, y=mean_rank, group=gene, col=term, text=Note)) + 
  facet_wrap(~population) + theme_minimal() + theme(legend.position="none"))  

ranks.adult.annot.ave$GOSlim_bin %>% unique()
```

