---
title: "02-Larval-data-analysis-QuantSeq2020"
author: "Laura H Spencer"
date: "5/22/2020"
output: html_document
---

### Load libraries and source scripts 

```{r, message=FALSE, warning=FALSE, results=FALSE}
source("../references/biostats.R")

list.of.packages <- c("DESeq2", "RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly") #add new libraries here 
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})
```
### Inspect total counts by sample for larval samples 
```{r}
load(file = "../results/gene-counts-filtered") #object = counts.filtered
load(file="../raw-data/key.filtered")

larval.samples <- key.filtered %>% filter(stage=="larvae") %>% dplyr::select(sample) %>% unlist() %>% as.vector() %>% as.character()

ggplotly(
ggplot(data = counts.filtered %>%
         dplyr::select(larval.samples) %>%
         colSums() %>% data.frame() %>%
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample"))  +
           geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample, larvae (pooled, whole-body)") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()))
```

### Load transformed counts object

```{r}
load(file = "../results/gene-counts-trans") #object = counts.t
```

### Extract larval samples only
```{r}
counts.t.larvae <- as.data.frame(counts.t[larval.samples, ])
counts.t.larvae[,1:3] #take a peak at the resulting dataframe 
```
## Optional 

### Pre-filtering - remove rows (genes) with less than a total of 10 reads (across all samples) (why?)

```{r}
keep <- colSums(counts.t.larvae) >= 100
counts.ts.larvae <- counts.t.larvae[,keep]
print(paste("# genes remaining after pre-filtering:", ncol(counts.ts.larvae)))
print(paste("# of genes dropped:", ncol(counts.t.larvae) - ncol(counts.ts.larvae), sep=" "))
```

### Remove other samples? 

```{r}
# For instance, if I wanted to remove all Oyster Bay C2 larvae 
# keep2 <- (key.filtered %>% filter(stage=="larvae" & population!="Oyster Bay C2"))$sample
# counts.ts.larvae <- counts.ts.larvae[keep2,]

# Or if I want to only compare 6A vs 10H
# key.filtered %>% filter(stage=="larvae" & (temp_pco2=="10_High" | temp_pco2=="6_Ambient")) %>% group_by(temp_pco2) %>% tally()
# keep2 <- (key.filtered %>% filter(stage=="larvae" & (temp_pco2=="10_High" | temp_pco2=="6_Ambient")))$sample
# counts.ts.larvae <- counts.ts.larvae[keep2,]
```


### How many genes were identified in each sample?

```{r}
ggplotly(
ggplot(data = data.frame(rowSums(counts.ts.larvae != 0)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample")) +
           geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total # genes by sample after pre-filtering, larvae") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```

## Use foa.plots to visualize data a bit: 

### In how many samples does each gene occur? 
  - The **first four plots portray the gene’ frequency of occurrence among samples** in a number of different ways – either as an empirical cumulative distribution function (ECDF) of gene occurrence or as a histogram of gene occurrence. 

### What is the mean abundance of each gene when it occurs (not averaging zeros for samples where it is absent)? 
  - The **fifth plot is an ECDF of gene mean abundance.** X-axis is samples, ranked from 1-n in terms of mean gene abundance. 
  
### Is the mean abundance of genes correlated with the number of samples they occur in? 
  - The **sixth plot is a scatter plot of frequency of occurrence against mean abundance**. Is there any apparent relationship between the two? Are the widespread genes also generally more abundant? Are there many widespread genes that occur at low abundance? Conversely, are there genes much less widespread, but abundant where they occur?

### Is the total abundance of gene in a sample correlated with the number of gene in a sample? To answer this question, first it is instructive to look at the number of gene per sample. 
  - The **eighth plot depicts the ECDF of sample richness.** Are there any interesting patterns? For example, do most samples support an average number of gene, while only a few samples supporting either very few or very many gene? Or is the pattern different?

### Second, what is the pattern in the distribution of sample total abundance? 
  - The **ninth plot is the ECDF of total sample abundance.** How does it compare to the ECDF of sample richness?

### Finally, to answer the question on the relation between total abundance and number of gene/sample ...
  - The **last plot is a scatter plot of the two variables.** Is there is relationship between the _number of genes per sample and the total abundance?_ Do gene-rich samples generally have a greater total abundance of those genes as well? 

```{r}
# note: you'll need to press return in the console for all plots 
#foa.plots(counts.ts.larvae)
```

### Merge sample key info to count data, then sort, and generate heat map for initial inspection by treatment 
```{r}
# merge count data with sample key, reset row names as sample names, and arrange by infection, then temperature, then day 

counts.tsk.larvae <- merge(x=key.filtered, by.x="sample", y=counts.ts.larvae, by.y="row.names") %>% 
  arrange(stage, population, pCO2.parent)  %>% column_to_rownames(var="sample")

head(counts.tsk.larvae[1:24]) #check out results of merge/arrange
#counts.tsk.larvae %>% dplyr::select(starts_with("OLUR")) #this is code to get only the gene columns 
```

### Generate heat map of counts before DESeq processing / analysis 

NOTE: scale="column" b/c range of counts is so huge, so counts have been scaled 

```{r}
pheatmap(data.matrix(counts.tsk.larvae %>% dplyr::select(starts_with("OLUR"))), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, 
                     show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE, 
                     scale="column", color=c("dodgerblue3", "goldenrod1"), 
                     main = "Oly Larvae Gene Counts", annotation_row=counts.tsk.larvae[,c("population", "pCO2.parent")],
         filename = "../results/heatmap-larval-counts.png")
```

Resulting heat map is located in results/ directory: 
![heatmap-larvae-counts.png](https://github.com/laurahspencer/O.lurida_QuantSeq-2020/blob/master/results/heatmap-larvae-counts.png?raw=true)

# Analysis in DESeq2  

### Reformat for DESeq, ensure correct sample order for 

NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

```{r}
all(rownames(counts.tsk.larvae) == counts.tsk.larvae %>% dplyr::select(starts_with("OLUR")) %>% t() %>% colnames()) #check that rownames of untransformed matrix match column names of transformed matrix. Should print 'TRUE' 
```

# Generate DESeq datasets with various treatment comparisons  

```{r}
#counts.DESeq <- counts.tsk.larvae[-which(rownames(counts.tsk.larvae) %in% "571_larvae"), grepl("OLUR", colnames(counts.tsk.larvae))] %>% t()
#key.DESeq <- counts.tsk.larvae[-which(rownames(counts.tsk.larvae) %in% "571_larvae"),c("population", "pCO2.parent")] 

dds.larvae <- DESeqDataSetFromMatrix(countData = counts.tsk.larvae[,grepl("OLUR", colnames(counts.tsk.larvae))] %>% t(),
                              colData = counts.tsk.larvae[,c("population", "temp.parent", "pCO2.parent", "week.collected", "prep.batch", "larval.day.collected")],  design = ~ population + temp.parent + pCO2.parent)
```

# Visualize data via PCAs and heat maps 

## Transform data 

- Here we transform counts using a variance stabilizing transformation (VST), since the rlog transformation threw an error and suggested using VST.  
- Here we use `blind=FALSE` b/c we are interested in differences explained by experimental design, and may wish to use this transformed data in downstream analyses. 
```{r}
vsd.larvae <- varianceStabilizingTransformation(dds.larvae, blind=FALSE)
```

## Visualize sample clustering via PCA (after transformation)

NOTE: Hover over points to see the sample numbers

```{r}
# PCA with points color coded by population 
ggplotly(
plotPCA(vsd.larvae, intgroup="population") + 
           ggtitle("PCA by population (var-stabilizing transformed)") + 
  geom_point(aes(size=vsd.larvae$larval.day.collected, text=colnames(vsd.larvae))) + theme_minimal() + stat_ellipse(), tooltip = "text")

# PCA with points color coded by parental temperature treatment
ggplotly(
plotPCA(vsd.larvae, intgroup="temp.parent") + 
           ggtitle("PCA by population (var-stabilizing transformed)") + 
  geom_point(size=3) + theme_minimal() + stat_ellipse(), tooltip = "text")

# PCA with points color coded by parental pCO2 exposure 
ggplotly(
plotPCA(vsd.larvae, intgroup="pCO2.parent") + 
           ggtitle("PCA by parental pCO2 exposure (var-stabilizing transformed)") + 
  geom_point(size=3, aes(text=colnames(vsd.larvae))) + theme_minimal() + stat_ellipse(), tooltip = "text")

# PCA with points color coded by temperature & parental pCO2 exposure 
ggplotly(
plotPCA(vsd.larvae, intgroup=c("temp.parent", "pCO2.parent")) + 
           ggtitle("PCA by parental pCO2 exposure (var-stabilizing transformed)") + 
  geom_point(size=3, aes(text=colnames(vsd.larvae))) + theme_minimal() + stat_ellipse(), tooltip = "text")

# PCA with points color coded by population, and pco2 factors 
ggplotly(
plotPCA(vsd.larvae, intgroup=c("population","pCO2.parent")) + 
           ggtitle("PCA by population + parental pCO2 (var-stabilizing transformed)") + 
  geom_point(size=3, aes(text=colnames(vsd.larvae))) + theme_minimal() + stat_ellipse(), tooltip = "text")

# PCA with points color coded by population, temperature, and pco2 factors 
ggplotly(
plotPCA(vsd.larvae, intgroup=c("population","temp.parent", "pCO2.parent")) + 
           ggtitle("PCA by population + parental pCO2 (var-stabilizing transformed)") + 
  geom_point(aes(size=vsd.larvae$larval.day.collected, text=colnames(vsd.larvae))) + theme_minimal() + stat_ellipse(), tooltip = "text")

# PCA with same 2 colors for all populations (for prelim. results presentation)
ggplotly(
  plotPCA(vsd.larvae, intgroup=c("population","pCO2.parent")) + 
           ggtitle("PCA by population + parental pCO2 (var-stabilizing transformed)") + 
    geom_point(size=4, aes(text=colnames(vsd.larvae))) + 
    scale_color_manual(values = c("#F8766D","#00BFC4","#F8766D","#00BFC4","#F8766D","#00BFC4","#F8766D","#00BFC4")) + stat_ellipse(), tooltip = "text")

# Check out other random effects 

# PCA with points color coded by library prep batch 
ggplotly(
plotPCA(vsd.larvae, intgroup=c("prep.batch")) + 
           ggtitle("PCA by population (var-stabilizing transformed)") + 
  geom_point(size=3) + theme_minimal() + stat_ellipse(), tooltip = "text")

# PCA with points color coded by time period colelcted (1, 2)
ggplotly(
plotPCA(vsd.larvae, intgroup=c("week.collected")) + 
           ggtitle("PCA by population (var-stabilizing transformed)") + 
  geom_point(size=3) + theme_minimal() + stat_ellipse(), tooltip = "text")

PCA.data.larvae <- plotPCA(vsd.larvae, intgroup=c("population","pCO2.parent"), returnData=TRUE)
save(PCA.data.larvae, file="../results/PCA.data.larvae")
```

### Generate heat maps before & after transformation  

```{r}
# extract treatment info from VSD transformation 
#vsd.larvae.df <- as.data.frame(colData(vsd.larvae)[,c("population", "pCO2.parent")])

# generate heatmap from untransformed counts 
#pheatmap(counts(dds.larvae), cluster_rows=FALSE, show_rownames=FALSE,
         #cluster_cols=T, annotation_col=vsd.larvae.df, scale = "row", main="QuantSeq, untransformed data (but scaled by rows")

# generate heatmap from VSD counts 
#pheatmap(assay(vsd.larvae), cluster_rows=FALSE, show_rownames=FALSE,
         #cluster_cols=T, annotation_col=vsd.larvae.df, main = "QuantSeq, VSD-transformed")
```

### Heatmap of the sample-to-sample distances
Another use of the transformed data is sample clustering. Here, we apply the dist function to the transpose of the transformed count matrix to get sample-to-sample distances.

A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.

```{r}
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

sampleDists.larvae <- dist(t(assay(vsd.larvae)))
sampleDistMatrix.larvae <- as.matrix(sampleDists.larvae)

# Here we show pCO2.parent + population
rownames(sampleDistMatrix.larvae) <- paste(vsd.larvae$population, vsd.larvae$pCO2.parent, sep="-") #set row names
colnames(sampleDistMatrix.larvae) <- NULL
pheatmap(sampleDistMatrix.larvae,
         clustering_distance_rows=sampleDists.larvae,
         clustering_distance_cols=sampleDists.larvae,
         col=rev(colors), fontsize = 6)
```

## Differential Expression Analysis - multifactor design 

### Run the function `DESeq` to assess differential expression 

```{r}
design(dds.larvae)
```


```{r}
dds.larvae.DESeq <- DESeq(dds.larvae) 
```
```{r}
# Larvae: Differential Gene Expression Analysis by _week collected_ (all populations combined)
# print("Comparison: period 1 vs. period 2")
# summary(results(dds.larvae.DESeq, contrast=c("week.collected", "1", "2"), alpha=0.05))
```

# Compare expression among parental pCO2, temperature, and population as sole factors 

## Larvae: Differential Gene Expression Analysis by _parental pCO2 exposure_ (all populations combined)

```{r}
print("Comparison: parental pCO2 - All populations")
summary(res.larvae.pco2 <- results(dds.larvae.DESeq, contrast=c("pCO2.parent", "Ambient", "High"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pCO2, larvae, comparison across all pops:",  sum(res.larvae.pco2$padj < 0.05, na.rm=TRUE))
# nope
```

## Larvae: Differential Gene Expression Analysis by _parental temperature exposure_ (all populations combined)

```{r}
levels(dds.larvae.DESeq$temp.parent)
print("Comparison: parental temperature - All populations")
summary(res.larvae.pco2 <- results(dds.larvae.DESeq, contrast=c("temp.parent", "6", "10"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by parent temperature, larvae, comparison across all pops:",  sum(res.larvae.pco2$padj < 0.05, na.rm=TRUE))
# nope
```

## Larvae: Differential Gene Expression Analysis by _population_ (all parental pCO2 exposure, combined)

```{r}
# Here are all the possible contrasts I can make 
levels(dds.larvae.DESeq$population)

print("Comparison: Dabob Bay vs. Fidalgo Bay")
summary(res.larvae.DBFB <- results(dds.larvae.DESeq, contrast=c("population", "Dabob Bay", "Fidalgo Bay"), alpha=0.05))

print("Comparison: Dabob Bay vs. Oyster Bay C1")
summary(res.larvae.DBOB1 <- results(dds.larvae.DESeq, contrast=c("population", "Dabob Bay", "Oyster Bay C1"), alpha=0.05))

print("Comparison: Dabob Bay vs. Oyster Bay C2")
summary(res.larvae.DBOB2 <- results(dds.larvae.DESeq, contrast=c("population", "Dabob Bay", "Oyster Bay C2"), alpha=0.05))

print("Comparison: Fidalgo Bay vs. Oyster Bay C1")
summary(res.larvae.FBOB1 <- results(dds.larvae.DESeq, contrast=c("population", "Fidalgo Bay", "Oyster Bay C1"), alpha=0.05))

print("Comparison: Fidalgo Bay vs. Oyster Bay C2")
summary(res.larvae.FBOB2 <- results(dds.larvae.DESeq, contrast=c("population", "Fidalgo Bay", "Oyster Bay C2"), alpha=0.05))

print("Comparison: Oyster Bay C1 vs. Oyster Bay C2")
summary(res.larvae.OB1OB2 <- results(dds.larvae.DESeq, contrast=c("population", "Oyster Bay C1", "Oyster Bay C2"), alpha=0.05))
```

## Count # of genes diff expressed  (p-value <0.05) in each population comparison 

```{r}
paste("No. of genes differentially expressed (padj<0.05) between Dabob & Fidalgo larvae:",  sum(res.larvae.DBFB$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between Dabob & Oyster Bay C1 larvae:",  sum(res.larvae.DBOB1$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between Dabob & Oyster Bay C2 larvae:",  sum(res.larvae.DBOB2$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between Fidalgo & Oyster Bay C1 larvae:",  sum(res.larvae.FBOB1$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between Fidalgo & Oyster Bay C2 larvae:",  sum(res.larvae.FBOB2$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) between Oyster Bay C1 & C2 larvae:",  sum(res.larvae.OB1OB2$padj < 0.05, na.rm=TRUE))
```

## Save population vs. population DEG lists for later comparisons 

```{r}
diffex.larvae.DBFB <- subset(res.larvae.DBFB, padj < 0.05)
diffex.larvae.DBOB1 <- subset(res.larvae.DBOB1, padj < 0.05)
diffex.larvae.DBOB2 <- subset(res.larvae.DBOB2, padj < 0.05)
diffex.larvae.FBOB1 <- subset(res.larvae.FBOB1, padj < 0.05)
diffex.larvae.FBOB2 <- subset(res.larvae.FBOB2, padj < 0.05)
# note: no differences between OB1 and OB2

# save all R objects to file for integration 
save(diffex.larvae.DBFB, file="../results/diffex.larvae.DBFB")
save(diffex.larvae.DBOB1, file="../results/diffex.larvae.DBOB1")
save(diffex.larvae.DBOB2, file="../results/diffex.larvae.DBOB2")
save(diffex.larvae.FBOB1, file="../results/diffex.larvae.FBOB1")
save(diffex.larvae.FBOB2, file="../results/diffex.larvae.FBOB2")

# merge and select unique genes for PCA 
degs.larvae.pops <- c(rownames(diffex.larvae.DBFB), 
  rownames(diffex.larvae.DBOB1),
  rownames(diffex.larvae.DBOB2),
  rownames(diffex.larvae.FBOB1),
  rownames(diffex.larvae.FBOB2)) %>% unique()
save(degs.larvae.pops, file="../results/degs.larvae.pops")
```

## Generate PCA with only differentially expressed genes among population 

This PCA represents the physiological differences among populations/cohorts of Olympia oyster larvae, upon maternal liberation. 

```{r}
# PCA with points color coded by tissue and pco2 factors 
#ggplotly(
  plotPCA(vsd.larvae[degs.larvae.pops,], intgroup=c("population")) + 
           ggtitle("Larval PCA by population, DEGs (var-stabilizing transformed)") + 
  geom_point(size=3, aes(text=colnames(vsd.larvae))) + theme_minimal() + stat_ellipse()#, tooltip = "text")
```

```{r}
# generate heatmap of population DEGs  
vsd.larvae.df <- as.data.frame(colData(vsd.larvae)[,c("population", "pCO2.parent")])

pheatmap(assay(vsd.larvae[degs.larvae.pops,]), cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=T, annotation_col=vsd.larvae.df[1], scale="row", main = "DEGs among Cohort, Larval Offspring",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3", `Oyster Bay C2` = "#e41a1c")))


# To not include OB2, use this instead: 
pheatmap(assay(vsd.larvae[degs.larvae.pops,rownames(vsd.larvae.df %>% filter(population!="Oyster Bay C2"))]), cluster_rows=T, show_rownames=FALSE, show_colnames=FALSE,
         cluster_cols=T, annotation_col=vsd.larvae.df[1], scale="row", main = "DEGs among Cohort, Larval Offspring",
         annotation_colors=list(population=c(`Dabob Bay` = "#4daf4a", `Fidalgo Bay` = "#377eb8", `Oyster Bay C1` = "#984ea3")))

#c("#4daf4a","#377eb8","#984ea3","#e41a1c")[vsd.larvae.df$population]
```

# Larvae: Differential Gene Expression Analysis by _parental temperature & pCO2 exposure_ (all populations combined)

```{r}
# Add a factor that defines interaction between pco2 and temperature 
dds.larvae.DESeq$temp.pCO2 <- factor(paste0(dds.larvae.DESeq$temp.parent, dds.larvae.DESeq$pCO2.parent))
design(dds.larvae.DESeq) <- ~ temp.pCO2 
dds.larvae.DESeq <- DESeq(dds.larvae.DESeq)
```

```{r}
levels(dds.larvae.DESeq$temp.pCO2)

print("Comparison: parental temperature & pCO2 - All populations")
summary(res.larvae.6 <- results(dds.larvae.DESeq, contrast=c("temp.pCO2", "6Ambient", "6High"), alpha=0.05))
summary(res.larvae.10 <- results(dds.larvae.DESeq, contrast=c("temp.pCO2", "10Ambient", "10High"), alpha=0.05))
summary(res.larvae.6a10h <- results(dds.larvae.DESeq, contrast=c("temp.pCO2", "6Ambient", "10High"), alpha=0.05))
summary(res.larvae.6h10a <- results(dds.larvae.DESeq, contrast=c("temp.pCO2", "6High", "10Ambient"), alpha=0.05))
# no DEGs! 
```

# Larvae: Differential Gene Expression Analysis by _Parental pCO2, within each population_ 

```{r}
# Add a factor that defines interaction group 
dds.larvae.DESeq$group <- factor(paste0(dds.larvae.DESeq$population, dds.larvae.DESeq$pCO2.parent))
design(dds.larvae.DESeq) <- ~ group 
dds.larvae.DESeq <- DESeq(dds.larvae.DESeq)
```

## Extract differential expression results / comparisons among different factors 

Create results objects, but summary of results are also shown

NOTE: can only compare two treatments at a time

```{r}
# Here are all the possible contrasts I can make 
levels(dds.larvae.DESeq$group)

print("Comparison: parental pCO2 - Fidalgo Bay")
summary(res.larvae.FB <- results(dds.larvae.DESeq, contrast=c("group", "Fidalgo BayHigh", "Fidalgo BayAmbient"), alpha=0.05))

print("Comparison: parental pCO2 - Dabob Bay")
summary(res.larvae.DB <- results(dds.larvae.DESeq, contrast=c("group", "Dabob BayHigh", "Dabob BayAmbient"), alpha=0.05))

print("Comparison: parental pCO2 - Oyster Bay Cohort 1")
summary(res.larvae.OB1 <- results(dds.larvae.DESeq, contrast=c("group", "Oyster Bay C1High", "Oyster Bay C1Ambient"), alpha=0.05))

print("Comparison: parental pCO2 - Oyster Bay Cohort 2")
summary(res.larvae.OB2 <- results(dds.larvae.DESeq, contrast=c("group", "Oyster Bay C2High", "Oyster Bay C2Ambient"), alpha=0.05))
```

## Count # of genes diff expressed  (p-value <0.05) in each comparison 

```{r}
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Fidalgo Bay larvae:",  sum(res.larvae.FB$padj < 0.05, na.rm=TRUE)) 
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Dabob Bay larvae:",  sum(res.larvae.DB$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Oyster Bay Cohort1 larvae:",  sum(res.larvae.OB1$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Oyster Bay Cohort2 larvae:",  sum(res.larvae.OB2$padj < 0.05, na.rm=TRUE))
```

## Extract stats for significantly different genes expressed in FB 

```{r}
diffex.larvae.FB <- subset(res.larvae.FB, padj < 0.05)
```

## Extract counts for differentially expressed genes for FB

```{r}
diffex.larvae.FB.counts <- subset(counts(dds.larvae.DESeq), rownames(dds.larvae.DESeq) %in% rownames(diffex.larvae.FB)) 
diffex.larvae.FB.counts <- diffex.larvae.FB.counts[,subset(key.filtered, stage=="larvae" & population=="Fidalgo Bay")$sample]
```

### Here we plot the genes differentially expressed in FB between parental pCO2 exposure  

```{r}
# Is the gene annotated? Nope. 
Olurida_gene_uniprot %>% filter(Name %in% rownames(diffex.larvae.FB)) %>%
  dplyr::select(contig, start, end, Name, Note, SPID)

a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.FB), intgroup=c("population", "pCO2.parent"), returnData = TRUE)

ggplot(subset(a, population=="Fidalgo Bay") %>% rownames_to_column("sample"),
       aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
  geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text() +
    theme_bw() +
    ggtitle(rownames(diffex.larvae.FB)) +
    theme(plot.title = element_text(hjust = 0.5))
```

```{r}
# Add a factor that defines interaction between population, temperature, and pco2 
dds.larvae.DESeq$pop.temp.pCO2 <- factor(paste0(dds.larvae.DESeq$population, dds.larvae.DESeq$temp.parent, dds.larvae.DESeq$pCO2.parent))
design(dds.larvae.DESeq) <- ~ pop.temp.pCO2 
dds.larvae.DESeq <- DESeq(dds.larvae.DESeq)
```

```{r}
levels(dds.larvae.DESeq$pop.temp.pCO2)

print("Comparison: parental temperature & pCO2 - within each population")
summary(res.larvae.FB6 <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Fidalgo Bay6Ambient", "Fidalgo Bay6High"), alpha=0.05))
summary(res.larvae.FB10 <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Fidalgo Bay10Ambient","Fidalgo Bay10High"), alpha=0.05))
summary(res.larvae.FB6a10h <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Fidalgo Bay6Ambient", "Fidalgo Bay10High"), alpha=0.05))
summary(res.larvae.FB6h10a <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Fidalgo Bay10Ambient", "Fidalgo Bay6High"), alpha=0.05)) # 2 DEGs 

summary(res.larvae.DB6 <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2","Dabob Bay6Ambient","Dabob Bay6High"), alpha=0.05))
summary(res.larvae.DB10 <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Dabob Bay10Ambient","Dabob Bay10High"), alpha=0.05))
summary(res.larvae.DB6a10h <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Dabob Bay6Ambient", "Dabob Bay10Ambient"), alpha=0.05)) # 1 DEG
summary(res.larvae.DB6h10a <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Dabob Bay6High", "Dabob Bay10Ambient"), alpha=0.05)) # 1 DEG 

summary(res.larvae.OB16 <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Oyster Bay C16Ambient","Oyster Bay C16High"), alpha=0.05)) # 2 DEGs 
summary(res.larvae.OB110 <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Oyster Bay C110Ambient","Oyster Bay C110High"), alpha=0.05)) # 3 DEGs 
summary(res.larvae.OB16a10h <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Oyster Bay C16Ambient", "Oyster Bay C110High"), alpha=0.05)) 
summary(res.larvae.OB16h10a <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Oyster Bay C16High", "Oyster Bay C110Ambient"), alpha=0.05))

summary(res.larvae.OB26 <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Oyster Bay C26Ambient","Oyster Bay C26High"), alpha=0.05))
summary(res.larvae.OB210 <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Oyster Bay C210Ambient","Oyster Bay C210High"), alpha=0.05))
summary(res.larvae.OB26a10h <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2",  "Oyster Bay C26Ambient", "Oyster Bay C210High"), alpha=0.05)) # 1 DEG
summary(res.larvae.OB26h10a <- results(dds.larvae.DESeq, contrast=c("pop.temp.pCO2", "Oyster Bay C26High", "Oyster Bay C210Ambient"), alpha=0.05))
```

```{r}
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, FB 6H vs. 10A larvae:",  sum(res.larvae.FB6h10a$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, DB 6A vs. 10H larvae:",  sum(res.larvae.DB6a10h$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, DB 6H vs. 10A larvae:",  sum(res.larvae.DB6h10a$padj < 0.05, na.rm=TRUE)) 
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, OB1 6A vs. 6H larvae:",  sum(res.larvae.OB16$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, OB1 10A vs. 10H larvae:",  sum(res.larvae.OB110$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, OB2 6A vs. 10H larvae:",  sum(res.larvae.OB26a10h$padj < 0.05, na.rm=TRUE))
```

## Extract stats for significantly different genes expressed in pop+temp+pco2 comparisons   

```{r}
diffex.larvae.FB6h10a <- subset(res.larvae.FB6h10a, padj < 0.05)
diffex.larvae.DB6a10h <- subset(res.larvae.DB6a10h, padj < 0.05)
diffex.larvae.DB6h10a <- subset(res.larvae.DB6h10a, padj < 0.05)
diffex.larvae.OB16 <- subset(res.larvae.OB16, padj < 0.05)
diffex.larvae.OB110 <- subset(res.larvae.OB110, padj < 0.05)
diffex.larvae.OB26a10h <- subset(res.larvae.OB26a10h, padj < 0.05)
```

## Extract counts for DEGs in pop+temp+pco2 comparisons   

```{r}
diffex.larvae.FB6h10a.counts <- subset(counts(dds.larvae.DESeq), rownames(dds.larvae.DESeq) %in% rownames(diffex.larvae.FB6h10a)) 
diffex.larvae.FB6h10a.counts <- diffex.larvae.FB6h10a.counts[,subset(key.filtered, stage=="larvae" & population=="Fidalgo Bay" & (temp_pco2=="6_High" | temp_pco2=="10_Ambient"))$sample, drop=FALSE]

diffex.larvae.DB6a10h.counts <- subset(counts(dds.larvae.DESeq), rownames(dds.larvae.DESeq) %in% rownames(diffex.larvae.DB6a10h)) 
diffex.larvae.DB6a10h.counts <- diffex.larvae.DB6a10h.counts[,subset(key.filtered, stage=="larvae" & population=="Dabob Bay" & (temp_pco2=="6_Ambient" | temp_pco2=="10_High"))$sample, drop=FALSE]

diffex.larvae.DB6h10a.counts <- subset(counts(dds.larvae.DESeq), rownames(dds.larvae.DESeq) %in% rownames(diffex.larvae.DB6h10a)) 
diffex.larvae.DB6h10a.counts <- diffex.larvae.DB6h10a.counts[,subset(key.filtered, stage=="larvae" & population=="Dabob Bay" & (temp_pco2=="6_High" | temp_pco2=="10_Ambient"))$sample, drop=FALSE]

diffex.larvae.OB16.counts <- subset(counts(dds.larvae.DESeq), rownames(dds.larvae.DESeq) %in% rownames(diffex.larvae.OB16)) 
diffex.larvae.OB16.counts <- diffex.larvae.OB16.counts[,subset(key.filtered, stage=="larvae" & population=="Oyster Bay C1" & temp.parent=="6")$sample, drop=FALSE]

diffex.larvae.OB110.counts <- subset(counts(dds.larvae.DESeq), rownames(dds.larvae.DESeq) %in% rownames(diffex.larvae.OB110)) 
diffex.larvae.OB110.counts <- diffex.larvae.OB110.counts[,subset(key.filtered, stage=="larvae" & population=="Oyster Bay C1" & temp.parent=="10")$sample, drop=FALSE]

diffex.larvae.OB26a10h.counts <- subset(counts(dds.larvae.DESeq), rownames(dds.larvae.DESeq) %in% rownames(diffex.larvae.OB26a10h)) 
diffex.larvae.OB26a10h.counts <- diffex.larvae.OB26a10h.counts[,subset(key.filtered, stage=="larvae" & population=="Oyster Bay C2" & (temp_pco2=="6_Ambient" | temp_pco2=="10_High"))$sample, drop=FALSE]
```

## Plot gene counts for DEGs in pop+temp+pco2 comparisons   

```{r}
# Are any genes annotated? A couple 
Olurida_gene_uniprot %>% filter(Name %in% 
  c(rownames(diffex.larvae.FB6h10a.counts),
  rownames(diffex.larvae.DB6a10h.counts),
  rownames(diffex.larvae.DB6h10a.counts),
  rownames(diffex.larvae.OB16.counts),
  rownames(diffex.larvae.OB110.counts),
  rownames(diffex.larvae.OB26a10h.counts))) %>% 
  dplyr::select(contig, start, end, Name, Note, SPID)

# FB 6-High vs. 10-Ambient
for (i in 1:nrow(diffex.larvae.FB6h10a.counts)) {
a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.FB6h10a.counts)[i], intgroup=c("pop.temp.pCO2"), returnData = TRUE)
print(ggplot(subset(a, pop.temp.pCO2=="Fidalgo Bay6High" | pop.temp.pCO2=="Fidalgo Bay10Ambient") %>% 
               rownames_to_column("sample"),
       aes(x=pop.temp.pCO2, y=count, color=pop.temp.pCO2, label=sample)) +
  geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text() +
    theme_bw() +
    ggtitle(rownames(diffex.larvae.FB6h10a.counts)[i]) +
    theme(plot.title = element_text(hjust = 0.5)))
}

# DB 6-Ambient vs. 10-High
for (i in 1:nrow(diffex.larvae.DB6a10h.counts)) {
a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.DB6a10h.counts)[i], intgroup=c("pop.temp.pCO2"), returnData = TRUE)
print(ggplot(subset(a, pop.temp.pCO2=="Dabob Bay6Ambient" | pop.temp.pCO2=="Dabob Bay10High") %>% 
               rownames_to_column("sample"),
       aes(x=pop.temp.pCO2, y=count, color=pop.temp.pCO2, label=sample)) +
  geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text() +
    theme_bw() +
    ggtitle(rownames(diffex.larvae.DB6a10h.counts)[i]) +
    theme(plot.title = element_text(hjust = 0.5)))
}

# DB 6-High vs. 10-Ambient
for (i in 1:nrow(diffex.larvae.DB6h10a.counts)) {
a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.DB6h10a.counts)[i], intgroup=c("pop.temp.pCO2"), returnData = TRUE)
print(ggplot(subset(a, pop.temp.pCO2=="Dabob Bay6High" | pop.temp.pCO2=="Dabob Bay10Ambient") %>% 
               rownames_to_column("sample"),
       aes(x=pop.temp.pCO2, y=count, color=pop.temp.pCO2, label=sample)) +
  geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text() +
    theme_bw() +
    ggtitle(rownames(diffex.larvae.DB6h10a.counts)[i]) +
    theme(plot.title = element_text(hjust = 0.5)))
}

# OB1 6-Ambient vs. 6-High
for (i in 1:nrow(diffex.larvae.OB16.counts)) {
a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.OB16.counts)[i], intgroup=c("pop.temp.pCO2"), returnData = TRUE)
print(ggplot(subset(a, pop.temp.pCO2=="Oyster Bay C16Ambient" | pop.temp.pCO2=="Oyster Bay C16High") %>% 
               rownames_to_column("sample"),
       aes(x=pop.temp.pCO2, y=count, color=pop.temp.pCO2, label=sample)) +
  geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text() +
    theme_bw() +
    ggtitle(rownames(diffex.larvae.OB16.counts)[i]) +
    theme(plot.title = element_text(hjust = 0.5)))
}

# OB1 10-Ambient vs. 10-High
for (i in 1:nrow(diffex.larvae.OB110.counts)) {
a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.OB110.counts)[i], intgroup=c("pop.temp.pCO2"), returnData = TRUE)
print(ggplot(subset(a, pop.temp.pCO2=="Oyster Bay C110Ambient" | pop.temp.pCO2=="Oyster Bay C110High") %>% 
               rownames_to_column("sample"),
       aes(x=pop.temp.pCO2, y=count, color=pop.temp.pCO2, label=sample)) +
  geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text() +
    theme_bw() +
    ggtitle(rownames(diffex.larvae.OB110.counts)[i]) +
    theme(plot.title = element_text(hjust = 0.5)))
}
 
# OB2 6-Ambient vs. 10-High
for (i in 1:nrow(diffex.larvae.OB26a10h.counts)) {
a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.OB26a10h.counts)[i], intgroup=c("pop.temp.pCO2"), returnData = TRUE)
print(ggplot(subset(a, pop.temp.pCO2=="Oyster Bay C26Ambient" | pop.temp.pCO2=="Oyster Bay C210High") %>% 
               rownames_to_column("sample"),
       aes(x=pop.temp.pCO2, y=count, color=pop.temp.pCO2, label=sample)) +
  geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text() +
    theme_bw() +
    ggtitle(rownames(diffex.larvae.OB26a10h.counts)[i]) +
    theme(plot.title = element_text(hjust = 0.5)))
}
```

# Remove:  OLUR_00010657, OLUR_00001136, OLUR_00023544 

# Are any of these genes annotated?

```{r}
diffex.larvae.inter.names <- c(rownames(diffex.larvae.FB6h10a),
rownames(diffex.larvae.DB6a10h),
rownames(diffex.larvae.DB6h10a),
rownames(diffex.larvae.OB16),
rownames(diffex.larvae.OB110),
rownames(diffex.larvae.OB26a10h))


#Is the gene annotated? Nope.
Olurida_gene_uniprot %>% filter(Name %in% diffex.larvae.inter.names) %>%
  dplyr::select(contig, start, end, Name, Note, SPID) %>%
  mutate(Protein=c(NA, NA, "Ras-related and estrogen-regulated growth inhibitor (RERG)", NA, NA, "Cholinesterase (BCHE)", NA, NA, NA, NA))

a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.2FB), intgroup=c("population", "pCO2.parent"), returnData = TRUE)

ggplot(subset(a, population=="Fidalgo Bay") %>% rownames_to_column("sample"),
       aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
  geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
    #geom_text() +
    theme_bw() +
    ggtitle(rownames(diffex.larvae.2FB)) +
    theme(plot.title = element_text(hjust = 0.5))
```

# Add Time Factor - Larvae: Differential Gene Expression Analysis by _Parental pCO2, within each population & time period_

```{r}
## Add a factor that defines interaction group 
# dds.larvae.DESeq$group2 <- factor(paste0(dds.larvae.DESeq$population, dds.larvae.DESeq$pCO2.parent, dds.larvae.DESeq$week.collected))
# design(dds.larvae.DESeq) <- ~ group2 
# dds.larvae.DESeq <- DESeq(dds.larvae.DESeq)
```

## Extract differential expression results / comparisons among different factors 

Create results objects, but summary of results are also shown

NOTE: can only compare two treatments at a time

```{r}
# # Here are all the possible contrasts I can make 
# levels(dds.larvae.DESeq$group2)
# 
# print("Comparison: parental pCO2 - Fidalgo Bay, Period 1 & Period 2")
# summary(res.larvae.1FB <- results(dds.larvae.DESeq, contrast=c("group2", "Fidalgo BayAmbient1", "Fidalgo BayHigh1"), alpha=0.05))
# summary(res.larvae.2FB <- results(dds.larvae.DESeq, contrast=c("group2", "Fidalgo BayHigh2", "Fidalgo BayAmbient2"), alpha=0.05))
# 
# print("Comparison: parental pCO2 - Dabob Bay, Period 1 & Period 2")
# summary(res.larvae.1DB <- results(dds.larvae.DESeq, contrast=c("group2", "Dabob BayHigh1", "Dabob BayAmbient1"), alpha=0.05))
# summary(res.larvae.2DB <- results(dds.larvae.DESeq, contrast=c("group2", "Dabob BayHigh2", "Dabob BayAmbient2"), alpha=0.05))
# 
# print("Comparison: parental pCO2 - Oyster Bay Cohort 1, Period 1 & Period 2")
# summary(res.larvae.1OB1 <- results(dds.larvae.DESeq, contrast=c("group2", "Oyster Bay C1High1", "Oyster Bay C1Ambient1"), alpha=0.05))
# summary(res.larvae.2OB1 <- results(dds.larvae.DESeq, contrast=c("group2", "Oyster Bay C1High2", "Oyster Bay C1Ambient2"), alpha=0.05))
# 
# print("Comparison: parental pCO2 - Oyster Bay Cohort 2, Period 1 & Period 2")
# summary(res.larvae.1OB2 <- results(dds.larvae.DESeq, contrast=c("group2", "Oyster Bay C2High1", "Oyster Bay C2Ambient1"), alpha=0.05))
# summary(res.larvae.2OB2 <- results(dds.larvae.DESeq, contrast=c("group2", "Oyster Bay C2High2", "Oyster Bay C2Ambient2"), alpha=0.05))
```

## Extract stats for significantly different genes expressed in FB  

```{r}
# diffex.larvae.2FB <- subset(res.larvae.2FB, padj < 0.05)
```

## Extract counts for differentially expressed genes for FB & DB cohort  

```{r}
# diffex.larvae.2FB.counts <- subset(counts(dds.larvae.DESeq), rownames(dds.larvae.DESeq) %in% rownames(diffex.larvae.2FB)) 
# diffex.larvae.2FB.counts <- diffex.larvae.2FB.counts[,subset(key.filtered, stage=="larvae" & population=="Fidalgo Bay")$sample]
```

## Plot gene counts for DEGs  

### Here we plot the genes differentially expressed in FB between parental pCO2 exposure  

```{r}
# Is the gene annotated? Nope. 
# Olurida_gene_uniprot %>% filter(Name %in% rownames(diffex.larvae.2FB)) %>%
#   select(contig, start, end, Name, Note, SPID)
# 
# a <-  plotCounts(dds.larvae.DESeq, gene=rownames(diffex.larvae.2FB), intgroup=c("population", "pCO2.parent"), returnData = TRUE)
# 
# ggplot(subset(a, population=="Fidalgo Bay") %>% rownames_to_column("sample"),
#        aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
#   geom_point(size=4, position=position_jitter(w = 0.2,h = 0)) +
#     #geom_text() +
#     theme_bw() +
#     ggtitle(rownames(diffex.larvae.2FB)) +
#     theme(plot.title = element_text(hjust = 0.5))
```


# Can I compare size to gene expression?

```{r}
load(file = "../results/larvae.size.mean")
larvae.size.mean
```

# Results for paper 

```{r}
# Within cohort pairwise comparisons among temperature & pCO2 treatments - DEG comparisons only 

paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, FB 6H vs. 10A larvae:",  sum(res.larvae.FB6h10a$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, DB 6A vs. 10H larvae:",  sum(res.larvae.DB6a10h$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, DB 6H vs. 10A larvae:",  sum(res.larvae.DB6h10a$padj < 0.05, na.rm=TRUE)) 
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, OB1 6A vs. 6H larvae:",  sum(res.larvae.OB16$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, OB1 10A vs. 10H larvae:",  sum(res.larvae.OB110$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, OB2 6A vs. 10H larvae:",  sum(res.larvae.OB26a10h$padj < 0.05, na.rm=TRUE))
```

# PCA with all genes using princomp and showing broken stick 

```{r}

# PCA using prcomp etc., but NOTE that I pull the covariance matrix from the SNPRelate PCA command  
pca.larvae.princomp.expr <- princomp(cov(assay(vsd.larvae))) #scale=F for variance-covariance matrix
pca.eigenval(pca.larvae.princomp.expr) #The Proporation of Variance = %variance 
pc.percent.larvae <- pca.eigenval(pca.larvae.princomp.expr)[2,1:4]*100
screeplot(pca.larvae.princomp.expr, bstick=TRUE) #looks like PC 1 & 2 sign.

tab.larvae.expr <- data.frame(sample.id = colnames(assay(vsd.larvae)),
    EV1.all = pca.larvae.princomp.expr$scores[,1],    # the first eigenvector
    EV2.all = pca.larvae.princomp.expr$scores[,2],    # the second eigenvector
    stringsAsFactors = FALSE)

tab.larvae.expr.annot <- left_join(tab.larvae.expr, sample.info.larvae, by=c("sample.id"="sample")) %>%
    mutate(population=as.factor(population), pCO2.parent=as.factor(pCO2.parent)) %>% drop_na(pCO2.parent) %>% droplevels()
#group.larvae.expr <- as.factor(tab.larvae.expr.annot$pop_code)

ggplot(tab.larvae.expr.annot, aes(x=EV1.all, y=EV2.all)) + geom_point(aes(shape=population, col=temp.parent:pCO2.parent), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Larval Expression, all genes, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.larvae[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.larvae[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_shape(name="Cohort") +
  scale_color_manual(values=c("#4575b4","#92c5de",  "#d73027","#f4a582"), name="Parental temp & pCO2 Exposure") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=temp.parent:pCO2.parent), size=0.3)

ggplot(tab.larvae.expr.annot, aes(x=EV1.all, y=EV2.all)) + geom_point(aes(col=population, shape=pCO2.parent), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Larval Expression, all genes, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.larvae[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.larvae[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(values=c("#4daf4a","#377eb8","#984ea3","#e41a1c"), name="Cohort") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + 
  stat_ellipse(aes(color=population), size=0.5)

# Plot PCA with points color coded by genetic PC axis 
ggplot(tab.larvae.expr.annot %>% left_join(tab.larvae.annot.gen[,c("EV1.gen", "EV2.gen", "sample.id")], by="sample.id"), 
       aes(x=EV1.all, y=EV2.all)) + geom_point(aes(col=EV2.gen, shape=population), size=3.5, alpha=0.8) + 
  theme_minimal() + ggtitle("Larval Expression, all genes, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.larvae[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.larvae[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_gradientn(colours = c("#e41a1c","#ff7f00","#377eb8"), name="Genetic PC1") +
  guides(colour = guide_legend(override.aes = list(size=3.5)))

save(tab.larvae.expr.annot, file="../results/tab.larvae.expr.annot")
```
# PCA with gene set that are DEGs among any pCO2 comparison (all pops, FB, DB, and OB1)

```{r}
pca.larvae.princomp.expr.co2 <-
  princomp(cov(assay(vsd.larvae[unique(c(
    rownames(diffex.larvae.DB6a10h),
    rownames(diffex.larvae.DB6h10a),
    rownames(diffex.larvae.OB16),
    rownames(diffex.larvae.OB110),
    rownames(diffex.larvae.OB26a10h))),])))
pca.eigenval(pca.larvae.princomp.expr.co2) #The Proporation of Variance = %variance 
pc.percent.larvae.pco2 <- pca.eigenval(pca.larvae.princomp.expr.co2)[2,1:4]*100
screeplot(pca.larvae.princomp.expr.co2, bstick=TRUE) #looks like PC 1

tab.larvae.expr.co2 <- data.frame(sample.id = colnames(assay(vsd.larvae)),
    EV1.pco2 = pca.larvae.princomp.expr.co2$scores[,1],    # the first eigenvector
    EV2.pco2 = pca.larvae.princomp.expr.co2$scores[,2],    # the second eigenvector
    stringsAsFactors = FALSE)

tab.larvae.expr.pco2.annot <- left_join(tab.larvae.expr.co2, sample.info.larvae, by=c("sample.id"="sample")) %>%
    mutate(population=as.factor(population), pCO2.parent=as.factor(pCO2.parent)) %>% drop_na(pCO2.parent) %>% droplevels()
#group.larvae.pco2.expr <- as.factor(tab.larvae.expr.pco2.annot$pop_code)

ggplot(tab.larvae.expr.pco2.annot, aes(x=EV1.pco2, y=EV2.pco2)) + geom_point(aes(shape=population, col=temp.parent:pCO2.parent), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Larval Expression, DEGs among parental Temp. & pCO2, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.larvae[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.larvae[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_shape(name="Cohort") +
  scale_color_manual(values=c("#4575b4","#92c5de",  "#d73027","#f4a582"), name="Parental temp & pCO2 Exposure") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=temp.parent:pCO2.parent), size=0.3)

# Plot PCA with points color coded by genetic PC axis 
ggplot(tab.larvae.expr.pco2.annot %>% left_join(tab.larvae.annot.gen[,c("EV1.gen", "EV2.gen", "sample.id")], by="sample.id"), 
       aes(x=EV1.pco2, y=EV2.pco2)) + geom_point(aes(col=EV2.gen, shape=population), size=3.5, alpha=0.8) + 
  theme_minimal() + ggtitle("Larval Expression,DEGs among parental Temp. & pCO2, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.larvae[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.larvae[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_gradientn(colours = c("#e41a1c","#377eb8","#ff7f00"), name="Genetic PC1") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) 

save(tab.larvae.expr.pco2.annot, file="../results/tab.larvae.expr.pco2.annot")
```
# PCA with gene set that are DEGs among any pairwise population comparison 

```{r}
pca.larvae.princomp.expr.pop <-
  princomp(cov(assay(vsd.larvae[degs.larvae.pops,])))
pca.eigenval(pca.larvae.princomp.expr.pop) #The Proporation of Variance = %variance 
pc.percent.larvae.pop <- pca.eigenval(pca.larvae.princomp.expr.pop)[2,1:4]*100
screeplot(pca.larvae.princomp.expr.pop, bstick=TRUE) #looks like PC 1, 2, & 3 are sign.

tab.larvae.expr.pop <- data.frame(sample.id = colnames(assay(vsd.larvae)),
    EV1.coh = pca.larvae.princomp.expr.pop$scores[,1],    # the first eigenvector
    EV2.coh = pca.larvae.princomp.expr.pop$scores[,2],    # the second eigenvector
    EV3.coh = pca.larvae.princomp.expr.pop$scores[,3],    # the third eigenvector
    stringsAsFactors = FALSE)

tab.larvae.expr.pop.annot <- left_join(tab.larvae.expr.pop, sample.info.larvae, by=c("sample.id"="sample"))  %>%
    mutate(population=as.factor(population), pCO2.parent=as.factor(pCO2.parent)) %>% drop_na(pCO2.parent) %>% droplevels()
group.larvae.pop.expr <- as.factor(tab.larvae.expr.pop.annot$pop_code)

ggplot(tab.larvae.expr.pop.annot, aes(x=EV1.coh, y=EV2.coh)) + geom_point(aes(shape=pCO2.parent, col=population), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Larval Expression PC1xPC2, DEGs among cohorts") + 
  xlab(paste("PC1 (", round(pc.percent.larvae.pop[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.larvae.pop[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(values=c("#4daf4a","#377eb8","#984ea3","#e41a1c"), name="Cohort") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=population), size=0.3)

ggplot(tab.larvae.expr.pop.annot, aes(x=EV1.coh, y=EV3.coh)) + geom_point(aes(shape=pCO2.parent, col=population), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Larval Expression PC1xPC3, DEGs among cohorts") + 
  xlab(paste("PC1 (", round(pc.percent.larvae.pop[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent.larvae.pop[3], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(values=c("#4daf4a","#377eb8","#984ea3","#e41a1c"), name="Cohort") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=population), size=0.3)

ggplot(tab.larvae.expr.pop.annot, aes(x=EV2.coh, y=EV3.coh)) + geom_point(aes(shape=pCO2.parent, col=population), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle("Larval Expression PC2xPC3, DEGs among cohorts") + 
  xlab(paste("PC2 (", round(pc.percent.larvae.pop[2], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent.larvae.pop[3], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_manual(values=c("#4daf4a","#377eb8","#984ea3","#e41a1c"), name="Cohort") +
  guides(colour = guide_legend(override.aes = list(size=3.5))) + stat_ellipse(aes(color=population), size=0.3)


# Plot PCA with points color coded by genetic PC axis 
ggplot(tab.larvae.expr.pop.annot %>% left_join(tab.larvae.annot.gen[,c("EV1.gen", "EV2.gen", "sample.id")], by="sample.id"), 
       aes(x=EV2.coh, y=EV3.coh)) + geom_point(aes(col=EV2.gen, shape=population), size=3.5, alpha=0.8) + 
  theme_minimal() + ggtitle("Larval Expression,DEGs among parental Temp. & pCO2, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent.larvae[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent.larvae[2], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  scale_color_gradientn(colours = c("red","blue"), name="Genetic PC1") +
  guides(colour = guide_legend(override.aes = list(size=3.5)))

save(tab.larvae.expr.pop.annot, file="../results/tab.larvae.expr.pop.annot")
```

# ==============
# Enrichment Analysis! using DEGs identified in  pairwise population comparisons

```{r}
# DEGs between DB & FB larvae 
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.larvae.DBFB)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between DB & OB1 larvae
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.larvae.DBOB1)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between FB & OB1
filter(Olurida_gene_uniprot, 
       Name %in% rownames(diffex.larvae.FBOB1)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# All genes for background
filter(Olurida_gene_uniprot, 
       Name %in% rownames(res.larvae.DBFB)) %>% 
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```
# For each GO term, assess whether counts were MORE abundant or LESS abundant in each cohort! 

```{r}
EnrBP.larvae.DBFB <- read_delim(file="../results/Larvae_cohort_DEGs_DBFB_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Dabob Bay\n~Fidalgo Bay", GO.category="Biological Processes")

EnrMF.larvae.DBFB <- read_delim(file="../results/Larvae_cohort_DEGs_DBFB_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Dabob Bay\n~Fidalgo Bay", GO.category="Molecular Function")

EnrBP.larvae.DBOB1 <- read_delim(file="../results/Larvae_cohort_DEGs_DBOB1_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Dabob Bay\n~Oyster Bay", GO.category="Biological Processes")

EnrMF.larvae.DBOB1 <- read_delim(file="../results/Larvae_cohort_DEGs_DBOB1_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Dabob Bay\n~Oyster Bay", GO.category="Molecular Function")

EnrBP.larvae.FBOB1 <- read_delim(file="../results/Larvae_cohort_DEGs_FBOB1_EnrichedBP.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Fidalgo Bay\n~Oyster Bay", GO.category="Biological Processes")

EnrMF.larvae.FBOB1 <- read_delim(file="../results/Larvae_cohort_DEGs_FBOB1_EnrichedMF.txt", delim = "\t") %>% 
  separate(into = c("GO", "process"), col=Term, "~") %>% clean_names() %>%
  dplyr::select(go, process, count, genes, fold_enrichment, p_value, fdr) %>% 
  mutate(perc_total=count/sum(count)) %>% 
  mutate(contrast="Fidalgo Bay\n~Oyster Bay", GO.category="Molecular Function")

all.go.enriched.cohorts.larvae <- rbind(
  EnrBP.larvae.DBFB,
  EnrMF.larvae.DBFB,
  EnrBP.larvae.DBOB1,
  EnrMF.larvae.DBOB1,
  EnrBP.larvae.FBOB1,
  EnrMF.larvae.FBOB1) %>% 
  left_join(GOslim, by=c("go"="GO_ID"))
save(all.go.enriched.cohorts.larvae, file = "../results/all.go.enriched.cohorts.larvae")

# Plot enriched Biological Processes 
all.go.enriched.cohorts.larvae %>% 
  mutate(go=as.factor(go)) %>% filter(GO.category=="Biological Processes") %>%
  ggplot(aes(y=str_wrap(process),x=contrast)) +  # NOTE: could instead use the next line to show GO Term ID's on the y-axis 
  geom_point(aes(alpha=fdr,size=count))+
 facet_wrap(~GO.category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.3), breaks = c(.25, .5, .75, 1), guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,12), breaks = c(2, 5, 10), guide = guide_legend(override.aes = list(col="gray50"))) + 
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), axis.text.x = element_text(size=7),
        legend.position = "right", axis.text.y=element_text(size=8.5), plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + ggtitle("Enriched function among cohorts, Larvae")

# Plot enriched Molecular Functions
all.go.enriched.cohorts.larvae %>% 
  mutate(go=as.factor(go)) %>% filter(GO.category=="Molecular Function") %>%
  ggplot(aes(y=str_wrap(process),x=contrast))+  # NOTE: could instead use the next line to show GO Term ID's on the y-axis 
  geom_point(aes(alpha=fdr,size=count)) +
 facet_wrap(~GO.category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.25), breaks = c(.25, .5, .75, 1), guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,12), breaks = c(2, 5, 10), guide = guide_legend(override.aes = list(col="gray50"))) + 
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), axis.text.x = element_text(size=7),
        legend.position = "right", axis.text.y=element_text(size=8.5), plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + ggtitle("Enriched function among cohorts, Larvae")
```

