---
title: "04b-Juvenile-data-analysis-deduplicated"
author: "Laura H Spencer"
date: "3/1/2021"
output: html_document
---


In this notebook I will import gene counts for juvenile samples that have been deduplicated. 

### Check working directory 

```{r}
getwd()
```

### Load libraries and source scripts 

```{r, message=FALSE, warning=FALSE, results=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
 
BiocManager::install("DESeq2")

source("../references/biostats.R")

list.of.packages <- c("DESeq2", "RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "corrplot", "PerformanceAnalytics") #add new libraries here 
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SNPRelate")
library(SNPRelate)
```

### Import QuantSeq sample info, library/file names, and then join 

```{r}
sample.info <- read.csv("../raw-data/quantseq2020_key.csv", header=T, na.strings = NA) %>%
          mutate_at(vars(lane, temp.parent), as.factor) %>% 
          mutate_at(vars(RNA.conc, endpoint.RFU, bioan.mean.bp), as.numeric) %>%
          mutate(sample_stage=paste(sample, stage, sep="_"))

filenames <- data.frame(read.table("../qc-processing/gatk/featureCounts-filenames.txt", header = F, stringsAsFactors = F, fill = FALSE)) %>%
          dplyr::rename(filename = 1) %>%
          mutate(sample = as.character(gsub("_.*","", filename))) %>% 
          left_join(sample.info[c("sample", "sample_stage")])

# Fill in NA value on the "Undetermined" row  
filenames[filenames$filename == "Undetermined","sample_stage"] <- "Undetermined"

(key <- full_join(sample.info, filenames) %>% filter(stage=="juvenile"))
```

### Import counts matrix file as dataframe. 
```{r}
counts.juv <- data.frame(read.table("../qc-processing/gatk/featurecounts-gene2kbdown.Rmatrix.txt", header = T, stringsAsFactors = F, fill = FALSE)) %>%
          column_to_rownames(var="Geneid") %>%
          rename_all(~as.character(filenames$sample_stage))
```

### Summarize counts and visualize (remove last column - that's undetermined counts)
```{r}
print(paste("Number of samples:", ncol(counts.juv), sep=" "))
print(paste("Total number of genes in dataframe:", prettyNum(nrow(counts.juv), big.mark = ","), sep=" "))
print(paste("Total counts, all samples:", prettyNum(sum(colSums(counts.juv)), big.mark = ","), sep=" "))
#print(paste("Counts for", colnames(counts.juv), ":",  prettyNum(colSums(counts.juv), big.mark = ","), sep=" "))

#inspect total counts by sample
#ggplotly(
  ggplot(data.frame(colSums(counts.juv)) %>% dplyr::rename(count.total = 1) %>% rownames_to_column(var="sample"))  + geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) #) 
```
### Transpose dataframe so each row = a sample (aka "objects"), and each column = genes (aka "variables") 
```{r}
counts.t.juv <- t(counts.juv) #transform data to have each sample a row, each column a gene 
```

## Optional 

### Pre-filtering - remove rows (genes) with less than a total of 10 reads (across all samples) (why?)

```{r}
keep <- colSums(counts.t.juv) >= 10
counts.ts.juv <- counts.t.juv[,keep]
print(paste("# genes remaining after pre-filtering:", ncol(counts.ts.juv)))
print(paste("# of genes dropped:", ncol(counts.t.juv) - ncol(counts.ts.juv), sep=" "))
```

### How many genes were identified in each sample?

```{r}
#ggplotly(
ggplot(data = data.frame(rowSums(counts.t.juv != 0)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample") %>% 
                  filter(grepl("_juvenile", sample))) +
           geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total # genes by sample, juvenile whole body") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())#) 
```

### Merge sample key info to count data, then sort, and generate heat map for initial inspection by treatment 
```{r}
# merge count data with sample key, reset row names as sample names, and arrange by infection, then temperature, then day 
counts.tsk.juv <- merge(x=key, by.x="sample_stage", y=counts.ts.juv, by.y="row.names") %>% 
  arrange(stage, population, pCO2.parent)  %>% column_to_rownames(var="sample_stage") 

head(counts.tsk.juv[1:24]) #check out results of merge/arrange
#counts.tsk.juv %>% dplyr::select(starts_with("OLUR")) #this is code to get only the gene columns 
```

### Generate heat map of counts before DESeq processing / analysis, save to file 

NOTE: scale="column" b/c range of counts is so huge, so counts have been scaled 

```{r}
pheatmap(data.matrix(counts.tsk.juv %>% dplyr::select(starts_with("OLUR"))), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, 
                     show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE, 
                     scale="column", color=c("dodgerblue3", "goldenrod1"), 
                     main = "Oly juvenile gene counts", annotation_row=counts.tsk.juv[,c("population", "pCO2.parent")],
         filename = "../results/heatmap-juvenile-counts.png")
```

Resulting heat map is located in results/ directory: 
![heatmap-juvenile-counts.png](https://github.com/laurahspencer/O.lurida_QuantSeq-2020/blob/master/results/heatmap-juvenile-counts.png?raw=true)

# Analysis in DESeq2  

### Reformat for DESeq, ensure correct sample order for 

NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

```{r}
all(rownames(counts.tsk.juv) == counts.tsk.juv %>% dplyr::select(starts_with("OLUR")) %>% t() %>% colnames()) #check that rownames of untransformed matrix match column names of transformed matrix. Should print 'TRUE' 
```

# Generate DESeq datasets with various treatment comparisons  

```{r}
#counts.DESeq <- counts.tsk.juv[-which(rownames(counts.tsk.juv) %in% "571_larvae"), grepl("OLUR", colnames(counts.tsk.juv))] %>% t()
#key.DESeq <- counts.tsk.juv[-which(rownames(counts.tsk.juv) %in% "571_larvae"),c("population", "pCO2.parent")] 

dds.juv <- DESeqDataSetFromMatrix(countData = counts.tsk.juv[,grepl("OLUR", colnames(counts.tsk.juv))] %>% t(),
                              colData = counts.tsk.juv[,c("population", "pCO2.parent")] ,
                              design = ~ population + pCO2.parent)
```

# Visualize data via PCAs and heat maps 

## Transform data 

- Here we transform counts using a variance stabilizing transformation (VST), since the rlog transformation threw an error and suggested using VST.  
- Here we use `blind=FALSE` b/c we are interested in differences explained by experimental design, and may wish to use this transformed data in downstream analyses. 
```{r}
vsd.juv <- varianceStabilizingTransformation(dds.juv, blind=FALSE)
```

## Visualize sample clustering via PCA (after transformation)

NOTE: Hover over points to see the sample numbers

```{r}
# PCA with points color coded by population 
#ggplotly(
plotPCA(vsd.juv, intgroup="population") + 
           ggtitle("PCA by population (var-stabilizing transformed)") + geom_point(size=3, aes(text=colnames(vsd.juv)))#, tooltip = "text")

# PCA with points color coded by parental pCO2 exposure 
#ggplotly(
plotPCA(vsd.juv, intgroup="pCO2.parent") + 
           ggtitle("PCA by parental pCO2 exposure (var-stabilizing transformed)") + geom_point(size=3, aes(text=colnames(vsd.juv)))#, tooltip = "text")

# PCA with points color coded by tissue and pco2 factors 
#ggplotly(
  plotPCA(vsd.juv, intgroup=c("population","pCO2.parent")) + 
           ggtitle("PCA by population + parental pCO2 (var-stabilizing transformed)") + geom_point(size=3, aes(text=colnames(vsd.juv)))#, tooltip = "text")
```

### Generate heat maps before & after transformation  

```{r}
# extract treatment info from VSD transformation 
vsd.juv.df <- as.data.frame(colData(vsd.juv)[,c("population", "pCO2.parent")])

# generate heatmap from untransformed counts 
pheatmap(counts(dds.juv), cluster_rows=FALSE, show_rownames=FALSE, 
         cluster_cols=T, annotation_col=vsd.juv.df, scale = "row", main="QuantSeq, untransformed data (but scaled by rows")

# generate heatmap from VSD counts 
pheatmap(assay(vsd.juv), cluster_rows=T, show_rownames=FALSE,
         cluster_cols=T, annotation_col=vsd.juv.df, main = "QuantSeq, VSD-transformed")
```

### Heatmap of the sample-to-sample distances
Another use of the transformed data is sample clustering. Here, we apply the dist function to the transpose of the transformed count matrix to get sample-to-sample distances.

A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.

```{r}
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

sampleDists <- dist(t(assay(vsd.juv)))
sampleDistMatrix <- as.matrix(sampleDists)

# Here we show pCO2.parent + population 
rownames(sampleDistMatrix) <- paste(vsd.juv$population, vsd.juv$pCO2.parent, sep="-") #set row names 
colnames(sampleDistMatrix) <- NULL
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=rev(colors), fontsize = 6)
```

## Differential Expression Analysis - multifactor design 

### Run the function `DESeq` to assess differential expression 

```{r}
dds.juv.DESeq <- DESeq(dds.juv) 
```

### Check out DESeq object design and treatments 

```{r}
colData(dds.juv.DESeq) # check out treatments 
design(dds.juv.DESeq)
```

# Juveniles: Differential Gene Expression Analysis by _parental pCO2 exposure_ (both populations combined)

```{r}
print("Comparison: parental pCO2 - All populations")
summary(res.juv.pco2 <- results(dds.juv.DESeq, contrast=c("pCO2.parent", "Ambient", "High"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pCO2, juvenile ctendia, comparison across both FB and DB pops:",  sum(res.juv.pco2$padj < 0.05, na.rm=TRUE))
print(diffex.juv <- subset(res.juv.pco2, padj < 0.05)) #extract stats for genes that are diff. expressed among parental pCO2, all pops combined 

diffex.juv.counts <- subset(counts(dds.juv.DESeq), rownames(dds.juv.DESeq) %in% rownames(diffex.juv)) 
diffex.juv.counts <- diffex.juv.counts[,subset(key, stage=="juvenile")$sample_stage]
```

## Heatmap of differentially expressed genes, both populations 

```{r}
dds.juv.df <- as.data.frame(colData(dds.juv)[,c("population", "pCO2.parent")]) 
dds.juv.df <- dds.juv.df[match(rownames(dds.juv.df), colnames(diffex.juv.counts)),]
all(colnames(diffex.juv.counts) == rownames(dds.juv.df)) #double check that samples are still in same order 

pheatmap(diffex.juv.counts, cluster_rows=T, show_rownames=T, cluster_columns=T, na.rm=TRUE, scale="row", main = "Gene Counts, all differentially expressed genes by parental pCO2, all juveniles", annotation_col=dds.juv.df[c(2)], color=rev(c("dodgerblue3", "goldenrod1")), fontsize = 7)
```

# Juveniles: Differential Gene Expression Analysis by _parental pCO2 exposure within population_ 

```{r}
# Add factor that defines interaction group 
dds.juv.DESeq$group <- factor(paste0(dds.juv.DESeq$population, dds.juv.DESeq$pCO2.parent))
design(dds.juv.DESeq) <- ~ group 
dds.juv.DESeq <- DESeq(dds.juv.DESeq)
```

## Extract differential expression results / comparisons among different factors 

Create results objects, but summary of results are also shown

NOTE: can only compare two treatments at a time
https://www.evernote.com/shard/s668/nl/155638310/b27aba76-366d-4047-b186-6951bbf472ac/
```{r}
# Here are all the possible contrasts I can make 
#levels(dds.multi.DESeq$group)

print("Comparison: parental pCO2 - Fidalgo Bay")
summary(res.juv.FB <- results(dds.juv.DESeq, contrast=c("group", "Fidalgo BayHigh", "Fidalgo BayAmbient"), alpha=0.05))

print("Comparison: parental pCO2 - Dabob Bay")
summary(res.juv.DB <- results(dds.juv.DESeq, contrast=c("group", "Dabob BayHigh", "Dabob BayAmbient"), alpha=0.05))
```

## Count # of genes diff expressed  (p-value <0.05) in each comparison 

```{r}
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Fidalgo Bay juvenile ctenidia:",  sum(res.juv.FB$padj < 0.05, na.rm=TRUE))
paste("No. of genes differentially expressed (padj<0.05) by parental pCO2, Dabob Bay juvenile ctenidia:",  sum(res.juv.DB$padj < 0.05, na.rm=TRUE))
```

## Extract stats for significantly different genes for each comparison 

```{r}
diffex.juv.FB <- subset(res.juv.FB, padj < 0.05)
diffex.juv.DB <- subset(res.juv.DB, padj < 0.05)
```

## Extract counts for differentially expressed genes for each comparison 

```{r}
diffex.juv.FB.counts <- subset(counts(dds.juv.DESeq), rownames(dds.juv.DESeq) %in% rownames(diffex.juv.FB)) 
diffex.juv.FB.counts <- diffex.juv.FB.counts[,subset(key, stage=="juvenile" & population=="Fidalgo Bay")$sample_stage]

diffex.juv.DB.counts <- subset(counts(dds.juv.DESeq), rownames(dds.juv.DESeq) %in% rownames(diffex.juv.DB))
diffex.juv.DB.counts <- diffex.juv.DB.counts[,subset(key, stage=="juvenile" & population=="Dabob Bay")$sample_stage]
```

# generate heatmap with differentially expressed genes within FB 

```{r}
dds.juv.df.FB <- as.data.frame(colData(dds.juv)[colData(dds.juv)$population=="Fidalgo Bay",c("population", "pCO2.parent")]) 
dds.juv.df.FB <- dds.juv.df.FB[match(rownames(dds.juv.df.FB), colnames(diffex.juv.FB.counts)),]
all(colnames(diffex.juv.FB.counts) == rownames(dds.juv.df.FB)) #double check that samples are still in same order 

pheatmap(diffex.juv.FB.counts, cluster_rows=T, show_rownames=FALSE, cluster_columns=T, na.rm=TRUE, scale="row", main = "FB Gene Counts, all differentially expressed genes among pCO2", annotation_col=dds.juv.df.FB[2], color=c("dodgerblue3", "goldenrod1"))
```

# generate heatmap with differentially expressed genes within DB 

```{r}
dds.juv.df.DB <- as.data.frame(colData(dds.juv)[colData(dds.juv)$population=="Dabob Bay",c("population", "pCO2.parent")]) 
dds.juv.df.DB <- dds.juv.df.DB[match(rownames(dds.juv.df.DB), colnames(diffex.juv.DB.counts)),]
all(colnames(diffex.juv.DB.counts) == rownames(dds.juv.df.DB)) #double check that samples are still in same order 

pheatmap(diffex.juv.DB.counts, cluster_rows=TRUE, show_rownames=FALSE, cluster_columns=T, na.rm=TRUE, scale="row", main = "DB Gene Counts, all differentially expressed genes among pCO2", annotation_col=dds.juv.df.DB[2], color=c("dodgerblue3", "goldenrod1"))
```
## Extract genes that are differentially expressed in Juveniles between parental pCO2 in both cohorts 

```{r}
Reduce(intersect, list(rownames(diffex.juv.FB),rownames(diffex.juv.DB)))
```

### Here we plot the 1 genes that's differentially expressed among parental pCO2 in both cohorts (OLUR_00005742)

```{r}
a <-  plotCounts(dds.juv.DESeq, gene="OLUR_00005742", intgroup=c("population", "pCO2.parent"), returnData = TRUE)
ggplot(a %>% rownames_to_column("sample"),
       aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
  #geom_boxplot() +
  geom_point(size=3, position=position_jitter(w = 0.1,h = 0)) +
    geom_text(size=3) +
    theme_bw() +
    ggtitle("OLUR_00005742 (unknown function)") +
    theme(plot.title = element_text(hjust = 0.5)) +
  facet_wrap(~population)
```

## Plot gene counts for differentially expressed genes 

It can also be useful to examine the counts of reads for a single gene across the groups. A simple function for making this plot is **`plotCounts`, which normalizes counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting.** The counts are grouped by the variables in intgroup, where more than one variable can be specified. Here we specify the gene which had the smallest p value from the results table created above. You can select the gene to plot by rowname or by numeric index.

### Here we plot the genes differentially expressed between FB cohorts whose parents were exposed to high pCO2 and not 

```{r}
juv.FB.p05.names <-rownames(diffex.juv.FB[order(diffex.juv.FB$padj),])
count_plots.juv.FB = list()
#plot the 37 genes with lowest p-values
for (i in 1:length(juv.FB.p05.names)) {
a <-  plotCounts(dds.juv.DESeq, gene=juv.FB.p05.names[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
b <- ggplot(subset(a, population=="Fidalgo Bay") %>% rownames_to_column("sample") %>% 
              mutate(sample=gsub("_juvenile", "", sample)),
       aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
  geom_point(position=position_jitter(w = 0.2,h = 0)) +
    #geom_text(size=3) +
    theme_bw() +
    ggtitle(juv.FB.p05.names[i]) +
    theme(plot.title = element_text(hjust = 0.5, size = 6), axis.text.x = element_blank(),
          axis.text.y = element_text(size=6),
          axis.title = element_blank(), axis.ticks.x = element_blank(), legend.position = "none")
count_plots.juv.FB[[i]] <- b
}

library(cowplot)
plot_grid(plotlist=count_plots.juv.FB) #plot them in a grid 
#count_plots.juv.FB #plot them all individually 
```
Identify location of genes to assess in IGV, write to bed file 

```{r}
write_delim(
  Olurida_gene_uniprot %>% filter(Name %in% juv.FB.p05.names) %>%
  select(contig, start, end, Name, Note, SPID),
  "../results/diffex.juvFB.bed", delim = '\t', col_names = F)

key %>% select(sample, population, pCO2.parent)
Olurida_gene_uniprot %>% filter(Name %in% juv.FB.p05.names) %>%
  select(contig, start, end, Name, Note, SPID)
```


### Here we plot the genes differentially expressed between DABOB BAY cohorts whose parents were exposed to high pCO2 and not 

```{r}
juv.DB.p05.names <-rownames(diffex.juv.DB[order(diffex.juv.DB$padj),])

count_plots.juv.DB = list()

for (i in 1:length(juv.DB.p05.names)) {
a <-  plotCounts(dds.juv.DESeq, gene=juv.DB.p05.names[i], intgroup=c("population", "pCO2.parent"), returnData = TRUE)
b <- ggplot(subset(a, population=="Dabob Bay") %>% rownames_to_column("sample") %>% 
              mutate(sample=gsub("_juvenile", "", sample)),
       aes(x=pCO2.parent, y=count, color=pCO2.parent, label=sample)) +
  geom_point(position=position_jitter(w = 0.2,h = 0)) +
    #geom_text(size=3) +
    theme_bw() +
    ggtitle(juv.DB.p05.names[i]) +
    theme(plot.title = element_text(hjust = 0.5, size = 6), axis.text.x = element_blank(),
          axis.text.y = element_text(size=6),
          axis.title = element_blank(), axis.ticks.x = element_blank(), legend.position = "none")
count_plots.juv.DB[[i]] <- b
}

plot_grid(plotlist=count_plots.juv.DB) #plot them in a grid 
#count_plots.juv.DB #plot them all individually 
```

```{r}
write_delim(
  Olurida_gene_uniprot %>% filter(Name %in% juv.DB.p05.names) %>%
  select(contig, start, end, Name, Note, SPID),
  "../results/diffex.juvDB.bed", delim = '\t', col_names = F)
```

