---
title: "QuantSeq-2020-Importing-Salmon-Counts"
author: "Laura H Spencer"
date: "5/13/2020"
output: html_document
---


### Load libraries

```{r message=FALSE, warning=FALSE, results=FALSE}
list.of.packages <- c("tidyverse", "reshape2", "here", "plotly", "tximport", "GenomicFeatures",  "readr") #add new libraries here 

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})
```

### Obtain session information

```{r}
sessionInfo()
getwd()
```





## Read in Salmon Count data 

Found [this helpful tutorial](https://www.hadriengourle.com/tutorials/rna/) that provides R script to load in and aggregate multiple Salmon output files. 

"Salmon did the quantifiation of the transcript level. We want to see which genes are differentially expressed, so we need to link the transcript names to the gene names. We can use our .gtf annotation for that, and the GenomicFeatures package:"

```{r}
txdb <- makeTxDbFromGFF("../references/20190709-Olurida_v081.stringtie.gtf")
txdb
keytypes(txdb)
print(k <- keys(txdb, keytype = "GENEID"))
print(k <- keys(txdb, keytype = "EXONID"))
print(k <- keys(txdb, keytype = "EXONNAME"))
txdb

tx2gene <- select(txdb, keys = k, keytype = "GENEID", columns = "TXNAME")
head(tx2gene)

```


