---
title: "DeSeq on QuantSeq"
output: html_document
---

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq2", version = "3.8")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("apeglm", version = "3.8")
library(ashr, verbose="F")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("vsn", version = "3.8")

list.of.packages <- c("DESeq2", "apeglm", "ashr", "pheatmap", "vsn") #add new libraries here 
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})

```

Import QuantSeq sample key

## Import count objects

```{r}
load("../results/counts.ts.adult.sv")
load("../results/counts.ts.juvenile.sv")
load("../results/counts.ts.larvae.sv")
```

From heatmap in notebook QuantSeq-multivariate-analysis, looks like samples X and Y had very few hits. Let's summarize gene counts by sample to confirm, and remove the samples with low gene counts if needed

```{r}
barplot(rowSums(counts.ts.adult.sv))
barplot(rowSums(counts.ts.juvenile.sv))
barplot(rowSums(counts.ts.larvae.sv))

remove.adult <- c()
remove.juvenile <- c()
remove.larvae <- c()

counts.adult <- counts.ts.adult.sv[!rownames(counts.ts.adult.sv) %in% "remove.adult", ] %>% t %>% as.data.frame()
counts.juvenile <- counts.ts.juvenile.sv[!rownames(counts.ts.juvenile.sv) %in% "remove.juvenile", ] %>% t() %>% as.data.frame()
counts.larvae <- counts.ts.larvae.sv[!rownames(counts.ts.larvae.sv) %in% "remove.larvae", ] %>% t() %>% as.data.frame()
```

###  Round all non-integers up, and remove genes that don't show up in any samples

```{r}
# counts.adult <- ceiling(counts.adult)
# counts.juvenile <- ceiling(counts.juvenile)
# counts.larvae <- ceiling(counts.larvae)
```

### Load key, reformat for DESeq, and filter to just include samples that I am analyzing 

NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

```{r}
load("../raw-data/key")

adult.key <- key %>%
          filter(stage=="adult") %>% 
          filter(sample_stage %in% colnames(counts.adult)) %>%
          select(sample_stage, sample, stage, temp.parent, pCO2.parent, population, pcr.cycles, prep.batch) %>%
          column_to_rownames('sample_stage')
all(rownames(adult.key) == colnames(counts.adult)) #confirm samples are in same order in count file and key 

juvenile.key <- key %>%
          filter(stage=="juvenile") %>% 
          filter(sample_stage %in% colnames(counts.juvenile)) %>%
          select(sample_stage, sample, stage, temp.parent, pCO2.parent, population, pcr.cycles, prep.batch) %>%
          column_to_rownames('sample_stage')
all(rownames(juvenile.key) == colnames(counts.juvenile)) #confirm samples are in same order in count file and key 

larvae.key <- key %>%
          filter(stage=="larvae") %>% 
          filter(sample_stage %in% colnames(counts.larvae)) %>%
          select(sample_stage, sample, stage, temp.parent, pCO2.parent, population, pcr.cycles, prep.batch) %>%
          column_to_rownames('sample_stage')
all(rownames(larvae.key) == colnames(counts.larvae)) #confirm samples are in same order in count file and key 
```

# Generate DESeq datasets with various treatment info 

```{r}
# Treatment = parental pCO2 
dds.adult.pCO2 <- DESeqDataSetFromMatrix(countData = counts.adult,
                              colData = adult.key,
                              design = ~ pCO2.parent)

# dds.juvenile.pCO2 <- DESeqDataSetFromMatrix(countData = counts.juvenile,
#                               colData = juvenile.key,
#                               design = ~ pCO2.parent)

dds.larvae.pCO2 <- DESeqDataSetFromMatrix(countData = counts.larvae,
                              colData = larvae.key,
                              design = ~ pCO2.parent)

# Treatment = population 
dds.adult.pop <- DESeqDataSetFromMatrix(countData = counts.adult,
                              colData = adult.key,
                              design = ~ population)

dds.juvenile.pop <- DESeqDataSetFromMatrix(countData = counts.juvenile,
                              colData = juvenile.key,
                              design = ~ population)

dds.larvae.pop <- DESeqDataSetFromMatrix(countData = counts.larvae,
                              colData = larvae.key,
                              design = ~ population)
```

## Start with the Larval pCO2 comparison 

Remove any rows (contigs) with zero hits

```{r}
keep <- rowSums(counts(dds.larvae.pCO2)) >= 1
dds.larvae.pCO2 <- dds.larvae.pCO2[keep,]
nrow(dds.larvae.pCO2)
colnames(dds.larvae.pCO2)
```

## Differential Expression Analysis

```{r}
dds.larvae.pCO2.DESeq <- DESeq(dds.larvae.pCO2)
```

```{r}
res.larvae.pCO2 <- results(dds.larvae.pCO2.DESeq)
```

### Log fold change shrinkage for visualization and ranking

```{r}
resultsNames(dds.larvae.pCO2.DESeq)[2] #extract name of coefficient
resLFC.larvae.pCO2 <- lfcShrink(dds.larvae.pCO2.DESeq, coef= resultsNames(dds.larvae.pCO2.DESeq)[2], type="apeglm")
```

```{r}
res.larvae.pCO2.ordered <- res.larvae.pCO2[order(res.larvae.pCO2$pvalue),]
summary(res.larvae.pCO2)
```

```{r}
sum(res.larvae.pCO2$padj < 0.1, na.rm=TRUE)
sum(res.larvae.pCO2$pvalue < 0.1, na.rm=TRUE)
```

```{r}
res05.larvae.pCO2 <- results(dds.larvae.pCO2.DESeq, alpha=0.05)
summary(res05.larvae.pCO2)
sum(res05.larvae.pCO2$padj < 0.05, na.rm=TRUE)
sum(res05.larvae.pCO2$pvalue < 0.05, na.rm=TRUE)
```

```{r}
plotMA(res.larvae.pCO2, ylim=c(-2,2))
```

```{r}
plotMA(resLFC.larvae.pCO2, ylim=c(-2,2))
```

```{r}
# because we are interested in treated vs untreated, we set 'coef=2'
resNorm.pH <- lfcShrink(dds.larvae.pCO2.DESeq, coef=2, type="normal")
resAsh.pH <- lfcShrink(dds.larvae.pCO2.DESeq, coef=2, type="ashr")
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC.larvae.pCO2, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm.pH, xlim=xlim, ylim=ylim, main="normal")
plotMA(resAsh.pH, xlim=xlim, ylim=ylim, main="ashr")
```

### plot gene counts for those with lowest p-values (below 0.05)

```{r}
pH.p05.names <-rownames(subset(res.larvae.pCO2.ordered, pvalue < 0.05)) #padj super high across the board - check out contigs with low pvalues instead ... 
for (i in 1:length(pH.p05.names)) {
          plotCounts(dds.larvae.pCO2.DESeq, gene=pH.p05.names[i], intgroup="pCO2.parent")
}
```


```{r}
vsd <- varianceStabilizingTransformation(dds.larvae.pCO2.DESeq, blind=FALSE)
rld <- rlog(dds.larvae.pCO2.DESeq, blind=FALSE)
ntd <- normTransform(dds.larvae.pCO2.DESeq)

meanSdPlot(assay(vsd))
meanSdPlot(assay(ntd))
meanSdPlot(assay(rld))

select <- order(rowMeans(counts(dds.larvae.pCO2.DESeq,normalized=TRUE)),
                decreasing=TRUE)
pH.df <- as.data.frame(colData(dds.larvae.pCO2.DESeq)[,c("pCO2.parent")])
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE)


plotPCA(ntd, intgroup=c("pCO2.parent"))
plotPCA(ntd, intgroup=c("pCO2.parent", "population"))
plotDispEsts(dds.larvae.pCO2.DESeq)
```

## Try larval population analysis 

```{r}
keep <- rowSums(counts(dds.larvae.pop)) >= 1
dds.larvae.pop <- dds.larvae.pop[keep,]
nrow(dds.larvae.pop)
colnames(dds.larvae.pop)

dds.larvae.pop.DESeq <- DESeq(dds.larvae.pop)

res.larvae.pop <- results(dds.larvae.pop.DESeq)

resultsNames(dds.larvae.pop.DESeq)[2] #extract name of coefficient
resLFC.larvae.pop <- lfcShrink(dds.larvae.pop.DESeq, coef= resultsNames(dds.larvae.pop.DESeq)[2], type="apeglm")

res.larvae.pop.ordered <- res.larvae.pop[order(res.larvae.pop$pvalue),]
summary(res.larvae.pop)

sum(res.larvae.pop$padj < 0.1, na.rm=TRUE)
sum(res.larvae.pop$pvalue < 0.1, na.rm=TRUE)

res05.larvae.pop <- results(dds.larvae.pop.DESeq, alpha=0.05)
summary(res05.larvae.pop)
sum(res05.larvae.pop$padj < 0.05, na.rm=TRUE)
sum(res05.larvae.pop$pvalue < 0.05, na.rm=TRUE)

plotMA(res.larvae.pop, ylim=c(-2,2))

plotMA(resLFC.larvae.pop, ylim=c(-2,2))

# because we are interested in treated vs untreated, we set 'coef=2'
resNorm.pH <- lfcShrink(dds.larvae.pop.DESeq, coef=2, type="normal")
resAsh.pH <- lfcShrink(dds.larvae.pop.DESeq, coef=2, type="ashr")
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC.larvae.pop, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm.pH, xlim=xlim, ylim=ylim, main="normal")
plotMA(resAsh.pH, xlim=xlim, ylim=ylim, main="ashr")

pH.p05.names <-rownames(subset(res.larvae.pop.ordered, pvalue < 0.05)) #padj super high across the board - check out contigs with low pvalues instead ... 
for (i in 1:length(pH.p05.names)) {
          plotCounts(dds.larvae.pop.DESeq, gene=pH.p05.names[i], intgroup="population")
}

vsd <- varianceStabilizingTransformation(dds.larvae.pop.DESeq, blind=FALSE)
rld <- rlog(dds.larvae.pop.DESeq, blind=FALSE)
ntd <- normTransform(dds.larvae.pop.DESeq)

meanSdPlot(assay(vsd))
meanSdPlot(assay(ntd))
meanSdPlot(assay(rld))

select <- order(rowMeans(counts(dds.larvae.pop.DESeq,normalized=TRUE)),
                decreasing=TRUE)
pH.df <- as.data.frame(colData(dds.larvae.pop.DESeq)[,c("population")])
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE)


plotPCA(vsd, intgroup=c("population"))
plotPCA(ntd, intgroup=c("population"))
plotPCA(rld, intgroup=c("population"))
plotDispEsts(dds.larvae.pop.DESeq)
```

























# Stopped here

Write out counts, and transformed data 

```{r}
pH.p05 <-as.data.frame(subset(res.larvae.pCO2.ordered, pvalue < 0.05))
write.csv(pH.p05, file="../results/DESeq/pH.p05.csv")

pH.p05.cts <- cts.gonad.rnd[rownames(cts.gonad.rnd) %in% pH.p05.names,]
write.csv(pH.p05.cts, file="../results/DESeq/pH.p05.counts.csv")

pH.p05.norm <- assay(rld[rownames(assay(rld)) %in% pH.p05.names,])
write.csv(pH.p05.norm, file="../results/DESeq/pH.p05.normalized.csv")

write.csv(pH.p05.names, file="../results/DESeq/pH.p05.contignames.csv")

write.csv(assay(rld), "../results/DESeq/pH.counts.csv")
```






# below is some extra script from the old QuantSeq notebook where I assessed larval data 

### plot gene counts for those with lowest p-values (below 0.05)

```{r}
print(pHLarvae.p05.names <-rownames(subset(resLarvae.pH.Ordered, pvalue < 0.05))) #padj super high across the board - check out contigs with low pvalues instead ... 
larvae.annot <- read.csv("results/DESeq/larvae.pH.p05-annot.csv", header=T, stringsAsFactors = FALSE)
#plot.axes <- c(0,0,0,0,0,0,7,8,9)

pdf(file = "results/DESeq/larval-diff-plots.pdf", width=5, height=6)
par(mfrow=(c(3,3)), mar=c(2,1,4,2), oma=c(1,4,4,2), col="gray25")

for (i in 1:length(pHLarvae.p05.names)) {
          y.lim <- c(0,max(assay(ddsLarvae.pH.DESeq)[rownames(assay(ddsLarvae.pH.DESeq)) == pHLarvae.p05.names[i],]))
          plotCounts(ddsLarvae.pH.DESeq, gene=pHLarvae.p05.names[i], intgroup="pH", col="gray40", bg=c("gray50", "steelblue")[ddsLarvae.pH.DESeq$pH], pch=21, cex=2, transform=FALSE, normalized=TRUE, ylim=c(y.lim*1.2), axes=FALSE, xlab = NA, main=larvae.annot[larvae.annot$Trinity.gene== pHLarvae.p05.names[i],"Gene"], col.main="gray40")
          axis(side=2,las=1,at=pretty(c(0,y.lim*1.1)))
          #if (i==plot.axes[i]) {
          #          axis(side=1)
          #}
}
mtext(side=2,text="Normalized counts", outer=T,line=2, col="gray40")
mtext(side=3,text="Larval gene counts by parental pH\nDifferentially expressed genes (top blast hit)", outer=T,line=-.5, col="gray30")
dev.off()
```


```{r}
vsd.larvae <- varianceStabilizingTransformation(ddsLarvae.pH.DESeq, blind=FALSE)
rld.larvae <- rlog(ddsLarvae.pH.DESeq, blind=FALSE)
ntd.larvae <- normTransform(ddsLarvae.pH.DESeq)

meanSdPlot(assay(vsd.larvae))
meanSdPlot(assay(ntd.larvae))
meanSdPlot(assay(rld.larvae))

select <- order(rowMeans(counts(ddsLarvae.pH.DESeq,normalized=TRUE)),
                decreasing=TRUE)
pHLarvae.df <- as.data.frame(colData(ddsLarvae.pH.DESeq)[,c("pH")])
pheatmap(assay(vsd.larvae)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=pHLarvae.df)

require(ggplot2)
plotPCA(vsd.larvae, intgroup=c("pH")) + geom_label(aes(label = name))
plotPCA(rld.larvae, intgroup=c("pH")) + geom_label(aes(label = name))
plotDispEsts(ddsLarvae.pH.DESeq)
View(key)
```


```{r}
Larvae.pH.p05 <-as.data.frame(subset(resLarvae.pH.Ordered, pvalue < 0.05))
write.csv(Larvae.pH.p05, file="../results/DESeq/larvae.pH.p05.csv")

Larvae.pH.p05.cts <- cts.larvae.rnd[rownames(cts.larvae.rnd) %in% pHLarvae.p05.names,]
write.csv(Larvae.pH.p05.cts, file="../results/DESeq/larvae.pH.p05.counts.csv")

write.csv(pHLarvae.p05.names, file="../results/DESeq/pHLarvae.p05.contignames.csv")

write.csv(assay(rld.larvae), file="results/DESeq/Larvae.pH.counts.csv")
```


## Check for outliers

```{r}
W <- res.larvae.pCO2$stat
maxCooks <- apply(assays(dds.larvae.pCO2.DESeq)[["cooks"]],1,max)
idx <- !is.na(W)
plot(rank(W[idx]), maxCooks[idx], xlab="rank of Wald statistic", ylab="maximum Cook's distance per gene", ylim=c(0,5), cex=.4, col=rgb(0,0,0,.3))
m <- ncol(dds.larvae.pCO2.DESeq)
p <- 3
abline(h=qf(.99, p, m - p))
```


# Differences between overwintering temp 

```{r}
keep <- rowSums(counts(dds.temp)) >= 1
dds.stage <- dds.temp[keep,]
nrow(dds.temp)
colnames(dds.temp)
dds.temp.DESeq <- DESeq(dds.temp)
res.temp <- results(dds.temp.DESeq)
res.temp.Ordered <- res.temp[order(res.temp$pvalue),]
summary(res.temp)
sum(res.temp$padj < 0.1, na.rm=TRUE)
sum(res.temp$pvalue < 0.05, na.rm=TRUE)
res05.temp <- results(dds.temp.DESeq, alpha=0.05)
summary(res05.temp)
sum(res05.temp$padj < 0.05, na.rm=TRUE)
sum(res05.temp$pvalue < 0.05, na.rm=TRUE)
temp.p05.names <-rownames(subset(res.temp.Ordered, pvalue < 0.05)) #padj super high across the board - check out contigs with low pvalues instead ... 
for (i in 1:length(temp.p05.names)) {
          plotCounts(dds.temp.DESeq, gene=temp.p05.names[i], intgroup="temperature")
}
```
